{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the process name to be human readable in htop\n",
    "import setproctitle\n",
    "setproctitle.setproctitle(\"11_Calculate_Disease_Enrichment\")\n",
    "\n",
    "from config import *\n",
    "from helper_functions import zip_res_range, unzip_res_range, odds_ratio, fasta2dict, flatten\n",
    "\n",
    "import pandas as pd\n",
    "pd.options.display.max_columns = 999\n",
    "\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "import networkx as nx\n",
    "\n",
    "from tqdm import tqdm, tqdm_notebook\n",
    "from tqdm._tqdm_notebook import tqdm_notebook\n",
    "\n",
    "tqdm.pandas(tqdm_notebook)\n",
    "tqdm_notebook.pandas()\n",
    "\n",
    "from collections import defaultdict\n",
    "\n",
    "%matplotlib inline\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook calculates the similarity log odds ratios for enrichment of disease mutations on human interactors of SARS-CoV-2 and on specific interfaces (either with SARS-CoV-2 proteins or with other human proteins). A summary of the most enriched disease terms in the human interactor set is retained (MedGen_Term_Enrichments.txt).\n",
    "\n",
    "\n",
    "- Inputs:\n",
    "  - Interface_Summary.txt\n",
    "  - Proteins.txt\n",
    "\n",
    "\n",
    "- Static Resource Dependencies:\n",
    "  - HGMD_201801_uniprotmapped.txt\n",
    "    - **NOTE:** Database publically available without, a blank header describing the layout is provided instead\n",
    "  - ClinVar2017_missense_uniprotmapped.txt\n",
    "  - H_sapiens_interfacesALL.txt\n",
    "  - keep_uniprot.fasta\n",
    "  - MGREL.RRF\n",
    "    - Downloaded on first run\n",
    "  - MGCONSO.RRF\n",
    "    - Downloaded on first run\n",
    "\n",
    "\n",
    "- Outputs:\n",
    "  - MedGen_Term_Enrichments.txt\n",
    "\n",
    "\n",
    "- Dependencies:\n",
    "  - Must be run after 08_Fetch_Top_Docks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parse Disease Mut Data from SARS2 Interactors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in Protein into to get list of human interactors\n",
    "proteins = pd.read_csv(\"{0}/Proteins.txt\".format(output_dir), sep=\"\\t\")\n",
    "uni2seq = proteins.set_index(\"ID\")[\"Sequence\"].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parse HGDM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in / subset HGMD dataset for only SARS2 interactors\n",
    "\n",
    "# NOTE: This script originally runs using a parsed version of HGMD created in the Yu Lab. Because\n",
    "#       HGMD requires users to register access their data (http://www.hgmd.cf.ac.uk/docs/register.html),\n",
    "#       this dataset cannot be shared through this repository.\n",
    "#\n",
    "#       Additionally, an HDMD Professional license must be purchased to access the most up to date version\n",
    "#       of the database (all data submitted in the past 3 years).\n",
    "#       \n",
    "#       For the purposes of this demo, a blank table including just the headers for this table are retained\n",
    "#       in the input.\n",
    "\n",
    "hgmd = pd.read_csv(\"{0}/HGMD_201801_uniprotmapped.txt\".format(resource_dir), sep=\"\\t\")\n",
    "hgmd_sub = hgmd[hgmd[\"UNIPROT_ACC\"].isin(uni2seq.keys())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>CLASS</th>\n",
       "      <th>DB</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_HG19</th>\n",
       "      <th>POS_HG38</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <th>DNA</th>\n",
       "      <th>CODON_CHANGE</th>\n",
       "      <th>PROT</th>\n",
       "      <th>UNIPROT_ACC</th>\n",
       "      <th>AA_POS</th>\n",
       "      <th>AA_CHANGE</th>\n",
       "      <th>PHEN</th>\n",
       "      <th>PMID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [ID, CLASS, DB, CHROM, POS_HG19, POS_HG38, REF, ALT, GENE, ENTREZ_GENE_ID, DNA, CODON_CHANGE, PROT, UNIPROT_ACC, AA_POS, AA_CHANGE, PHEN, PMID]\n",
       "Index: []"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print len(hgmd_sub)\n",
    "hgmd_sub.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Approximate fraction of all human proteins with disease annotations\n",
    "print len(hgmd[\"UNIPROT_ACC\"].unique()) / 20000.0\n",
    "\n",
    "# Fraction of SARS2 interactors with disease annotations\n",
    "len(hgmd_sub[\"UNIPROT_ACC\"].value_counts()) / float(len([x for x in uni2seq if not \"COVID\" in x]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             Case  Non-Case\n",
      "Exposed       0.0      10.0\n",
      "Non-Exposed   0.0   20340.0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(nan, nan, nan, nan)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Simple Odds Ratio for enrichment of general disease status\n",
    "# in SARS2 interactor sub-set vs. whole proteome\n",
    "n_proteome = 20350\n",
    "n_interactors = len([x for x in uni2seq if not \"COVID\" in x])\n",
    "\n",
    "a = len(hgmd_sub[\"UNIPROT_ACC\"].unique())\n",
    "b = n_interactors - len(hgmd_sub[\"UNIPROT_ACC\"].unique())\n",
    "c = len(hgmd[\"UNIPROT_ACC\"].unique()) - len(hgmd_sub[\"UNIPROT_ACC\"].unique())\n",
    "d = n_proteome - (a + b + c)\n",
    "\n",
    "exposure_mask = np.zeros(a + b + c + d)\n",
    "case_mask = np.zeros(a + b + c + d)\n",
    "exposure_mask[:a+b] = 1\n",
    "case_mask[:a] = 1\n",
    "case_mask[a+b:a+b+c] = 1\n",
    "\n",
    "odds_ratio(exposure_mask, case_mask, log_odds=True, verbose=True, two_sided=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "48a34ff97ce046068b61938f953d2ff5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=1, bar_style=u'info', max=1), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Map phenotype to number of genes associated with it for each\n",
    "# the whole proteome and the subset containing SARS2 interactors\n",
    "phen2genes_hgmd = defaultdict(int)\n",
    "phen2genes_hgmd_sub = defaultdict(int)\n",
    "keep_unis = set(uni2seq.keys())\n",
    "for g in tqdm_notebook(hgmd.drop_duplicates([\"UNIPROT_ACC\", \"PHEN\"]).groupby(\"UNIPROT_ACC\")):\n",
    "    for phen in g[1][\"PHEN\"]:\n",
    "        for real_phen in phen.split(\";\"):\n",
    "            phen2genes_hgmd[real_phen] += 1\n",
    "            if(g[0] in keep_unis):\n",
    "                phen2genes_hgmd_sub[real_phen] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "641e509d50da481fb710d448af23d514",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=1, bar_style=u'info', max=1), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Calculate Odds ratio per disease term\n",
    "phen2OR_hgmd = dict()\n",
    "for phen in tqdm_notebook(phen2genes_hgmd.keys()):\n",
    "    # Case = Gene has phenotype vs. Gene does not have phenotype\n",
    "    # Exposure = Gene in interactome vs. Gene no in interactome\n",
    "    a = phen2genes_hgmd_sub[phen]\n",
    "    b = n_interactors - phen2genes_hgmd_sub[phen]\n",
    "    c = phen2genes_hgmd[phen] - phen2genes_hgmd_sub[phen]\n",
    "    d = n_proteome - (a + b + c)\n",
    "    \n",
    "    #if(0 in [a, b, c, d]):\n",
    "    #    continue\n",
    "    \n",
    "    exposure_mask = np.zeros(a + b + c + d)\n",
    "    case_mask = np.zeros(a + b + c + d)\n",
    "    exposure_mask[:a+b] = 1\n",
    "    case_mask[:a] = 1\n",
    "    case_mask[(a+b):(a+b+c)] = 1\n",
    "    \n",
    "    OR, up, low, p = odds_ratio(exposure_mask, case_mask, log_odds=True)\n",
    "    \n",
    "    phen2OR_hgmd[phen] = [OR, up, low, p, a, b, c, d]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Concatenate summary for odds ratio from all disease terms\n",
    "# Counts Column Interpretations...\n",
    "# A - Has gene phenotype and is a SARS2 interactor\n",
    "# B - No gene phenotype but is a SARS2 interactor\n",
    "# C - Has gene phenotype but is not a SARS2 interactor\n",
    "# D - No gene phenotype and not a SARS2 interactor\n",
    "OR_summary_hgmd = pd.DataFrame([[k] + v for k, v in phen2OR_hgmd.iteritems()], columns=[\"Phen\", \"OR\", \"Lower\", \"Upper\", \"p-value\", \"A\", \"B\", \"C\", \"D\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phen</th>\n",
       "      <th>OR</th>\n",
       "      <th>Lower</th>\n",
       "      <th>Upper</th>\n",
       "      <th>p-value</th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Phen, OR, Lower, Upper, p-value, A, B, C, D]\n",
       "Index: []"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check to see if there are any terms with perfect enrichment\n",
    "# (all disease assocated genes are in interactome)\n",
    "# Particularly is the A count is high\n",
    "# NOTE: The odds ratio for this case can't be calculated, but\n",
    "#       may be interesting, so we look at it separately\n",
    "OR_summary_hgmd[OR_summary_hgmd[\"C\"] == 0].sort_values(\"A\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phen</th>\n",
       "      <th>OR</th>\n",
       "      <th>Lower</th>\n",
       "      <th>Upper</th>\n",
       "      <th>p-value</th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Phen, OR, Lower, Upper, p-value, A, B, C, D]\n",
       "Index: []"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Look at significantly enriched terms\n",
    "OR_summary_hgmd[(OR_summary_hgmd[\"p-value\"] <= 0.05) & (OR_summary_hgmd[\"OR\"] > 0)].sort_values(\"A\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parse ClinVar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in / subset ClinVar dataset for only SARS2 interactors\n",
    "\n",
    "# NOTE: This script runs using a parsed version of the ClinVar Database created in the Yu Lab.\n",
    "#       The original downloading and parsing was performed by Siqi (Charles) Liang. Since the\n",
    "#       ClinVar database is publically available (https://ftp.ncbi.nlm.nih.gov/pub/clinvar/)\n",
    "#       the parsed version of the databse is provided.\n",
    "#       \n",
    "#       The enrichment results from this demo still aren't particularly informative because such\n",
    "#       a small subset of interactions is selected for the demo. But the method of calculating the\n",
    "#       enrichment should at least be clear.\n",
    "\n",
    "clinvar = pd.read_csv(\"{0}/ClinVar2017_missense_uniprotmapped.txt\".format(resource_dir), sep=\"\\t\")\n",
    "clinvar_sub = clinvar[clinvar[\"UNIPROT_ACC\"].map(lambda x: x in uni2seq.keys())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.23735\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.3"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Approximate fraction of all human proteins with disease annotations\n",
    "print len(clinvar[\"UNIPROT_ACC\"].unique()) / 20000.0\n",
    "\n",
    "# Fraction of SARS2 interactors with disease annotations\n",
    "len(clinvar_sub[\"UNIPROT_ACC\"].value_counts()) / float(len([x for x in uni2seq if not \"COVID\" in x]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "               Case  Non-Case\n",
      "Exposed         3.0       7.0\n",
      "Non-Exposed  4744.0   15596.0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(0.49460772351518417,\n",
       " 2.1326212146975201,\n",
       " -1.1434057676671521,\n",
       " 0.30970978247789138)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Simple Odds Ratio for enrichment of general disease status\n",
    "# in SARS2 interactor sub-set vs. whole proteome\n",
    "n_proteome = 20350\n",
    "n_interactors = len([x for x in uni2seq if not \"COVID\" in x])\n",
    "\n",
    "a = len(clinvar_sub[\"UNIPROT_ACC\"].unique())\n",
    "b = n_interactors - len(clinvar_sub[\"UNIPROT_ACC\"].unique())\n",
    "c = len(clinvar[\"UNIPROT_ACC\"].unique()) - len(clinvar_sub[\"UNIPROT_ACC\"].unique())\n",
    "d = n_proteome - (a + b + c)\n",
    "\n",
    "exposure_mask = np.zeros(a + b + c + d)\n",
    "case_mask = np.zeros(a + b + c + d)\n",
    "exposure_mask[:a+b] = 1\n",
    "case_mask[:a] = 1\n",
    "case_mask[a+b:a+b+c] = 1\n",
    "\n",
    "odds_ratio(exposure_mask, case_mask, log_odds=True, verbose=True, two_sided=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e79fa40f076f4754b167cdb6a1dd4a8c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=4747), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Map phenotype to number of genes associated with it for each\n",
    "# the whole proteome and the subset containing SARS2 interactors\n",
    "phen2genes_clinvar = defaultdict(int)\n",
    "phen2genes_clinvar_sub = defaultdict(int)\n",
    "keep_unis = set(uni2seq.keys())\n",
    "for g in tqdm_notebook(clinvar.drop_duplicates([\"UNIPROT_ACC\", \"PHEN\"]).groupby(\"UNIPROT_ACC\")):\n",
    "    for phen in g[1][\"PHEN\"]:\n",
    "        phen = str(phen)\n",
    "        for real_phen in phen.split(\";\"):\n",
    "            phen2genes_clinvar[real_phen] += 1\n",
    "            if(g[0] in keep_unis):\n",
    "                phen2genes_clinvar_sub[real_phen] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3415defdf93c4dc6b9047e899f7b0693",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=6241), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Calculate Odds ratio per disease term\n",
    "phen2OR_clinvar = dict()\n",
    "for phen in tqdm_notebook(phen2genes_clinvar.keys()):\n",
    "    # Case = Gene has phenotype vs. Gene does not have phenotype\n",
    "    # Exposure = Gene in interactome vs. Gene no in interactome\n",
    "    a = phen2genes_clinvar_sub[phen]\n",
    "    b = n_interactors - phen2genes_clinvar_sub[phen]\n",
    "    c = phen2genes_clinvar[phen] - phen2genes_clinvar_sub[phen]\n",
    "    d = n_proteome - (a + b + c)\n",
    "    \n",
    "    #if(0 in [a, b, c, d]):\n",
    "    #    continue\n",
    "    \n",
    "    exposure_mask = np.zeros(a + b + c + d)\n",
    "    case_mask = np.zeros(a + b + c + d)\n",
    "    exposure_mask[:a+b] = 1\n",
    "    case_mask[:a] = 1\n",
    "    case_mask[(a+b):(a+b+c)] = 1\n",
    "    \n",
    "    OR, up, low, p = odds_ratio(exposure_mask, case_mask, log_odds=True)\n",
    "    \n",
    "    phen2OR_clinvar[phen] = [OR, up, low, p, a, b, c, d]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Concatenate summary for odds ratio from all disease terms\n",
    "# Counts Column Interpretations...\n",
    "# A - Has gene phenotype and is a SARS2 interactor\n",
    "# B - No gene phenotype but is a SARS2 interactor\n",
    "# C - Has gene phenotype but is not a SARS2 interactor\n",
    "# D - No gene phenotype and not a SARS2 interactor\n",
    "OR_summary_clinvar = pd.DataFrame([[k] + v for k, v in phen2OR_clinvar.iteritems()], columns=[\"Phen\", \"OR\", \"Lower\", \"Upper\", \"p-value\", \"A\", \"B\", \"C\", \"D\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phen</th>\n",
       "      <th>OR</th>\n",
       "      <th>Lower</th>\n",
       "      <th>Upper</th>\n",
       "      <th>p-value</th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Impdh2 enzyme activity, variation in</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>20340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>Hyperphosphatasia with mental retardation synd...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>20340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3439</th>\n",
       "      <td>nicotine response - Efficacy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>20340</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Phen  OR  Lower  Upper  \\\n",
       "9                  Impdh2 enzyme activity, variation in NaN    NaN    NaN   \n",
       "566   Hyperphosphatasia with mental retardation synd... NaN    NaN    NaN   \n",
       "3439                       nicotine response - Efficacy NaN    NaN    NaN   \n",
       "\n",
       "      p-value  A  B  C      D  \n",
       "9         NaN  1  9  0  20340  \n",
       "566       NaN  1  9  0  20340  \n",
       "3439      NaN  1  9  0  20340  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check to see if there are any terms with perfect enrichment\n",
    "# (all disease assocated genes are in interactome)\n",
    "# Particularly is the A count is high\n",
    "# NOTE: The odds ratio for this case can't be calculated, but\n",
    "#       may be interesting, so we look at it separately\n",
    "OR_summary_clinvar[OR_summary_clinvar[\"C\"] == 0].sort_values(\"A\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phen</th>\n",
       "      <th>OR</th>\n",
       "      <th>Lower</th>\n",
       "      <th>Upper</th>\n",
       "      <th>p-value</th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4982</th>\n",
       "      <td>Hyperphosphatasia with Intellectual Disability...</td>\n",
       "      <td>11.311890</td>\n",
       "      <td>14.311111</td>\n",
       "      <td>8.312670</td>\n",
       "      <td>1.443290e-13</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>20338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2667</th>\n",
       "      <td>nan</td>\n",
       "      <td>3.613368</td>\n",
       "      <td>6.601339</td>\n",
       "      <td>0.625398</td>\n",
       "      <td>1.777873e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>183</td>\n",
       "      <td>20157</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Phen         OR      Lower  \\\n",
       "4982  Hyperphosphatasia with Intellectual Disability...  11.311890  14.311111   \n",
       "2667                                                nan   3.613368   6.601339   \n",
       "\n",
       "         Upper       p-value  A  B    C      D  \n",
       "4982  8.312670  1.443290e-13  2  8    2  20338  \n",
       "2667  0.625398  1.777873e-02  1  9  183  20157  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Look at significantly enriched terms\n",
    "OR_summary_clinvar[(OR_summary_clinvar[\"p-value\"] <= 0.05) & (OR_summary_clinvar[\"OR\"] > 0)].sort_values(\"A\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate Disease Enrichment at Interfaces"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up masks to use for calculating enrichment\n",
    "\n",
    "# Mask for interface residues based on viral interfaces\n",
    "uni2ires_mask_viral = dict()\n",
    "\n",
    "# Mask for interface residues based on other human-human interfaces\n",
    "uni2ires_mask_insider = dict()\n",
    "\n",
    "# Mask for locations of HGMD disease variants\n",
    "uni2hgmd_mask = dict()\n",
    "\n",
    "# Mask for locations of CLinVar disease variants\n",
    "uni2clinvar_mask = dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in viral interface info\n",
    "ires_df = pd.read_csv(\"{0}/Interface_Summary.txt\".format(output_dir), sep=\"\\t\")\n",
    "ires_df = ires_df.sort_values([\"P1\", \"P2\", \"Source\"]).drop_duplicates([\"P1\", \"P2\"])\n",
    "#ires_df = ires_df[ires_df[\"Source\"] == \"ECLAIR\"]\n",
    "#ires_df = ires_df[ires_df[\"Source\"] == \"Docking\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e1eccbf98c0540edb227def98dc72d75",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=10), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Populate Ires Viral Masks\n",
    "for p2, p2_ires in tqdm_notebook(ires_df[[\"P2\", \"P2_Ires\"]].values):\n",
    "    # If no mask exists for this protein, initialize an all-zeros mask\n",
    "    if(not p2 in uni2ires_mask_viral):\n",
    "        uni2ires_mask_viral[p2] = np.zeros(len(uni2seq[p2]))\n",
    "    \n",
    "    if(not pd.isnull(p2_ires)):\n",
    "        # Parse interface positions\n",
    "        p2_ires = [int(x)-1 for x in p2_ires.split(\",\")]\n",
    "        \n",
    "        # Update Mask\n",
    "        uni2ires_mask_viral[p2][p2_ires] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in INSIDER human-human interface annotations\n",
    "insider = pd.read_csv(\"{0}/H_sapiens_interfacesALL.txt\".format(resource_dir), sep=\"\\t\")\n",
    "\n",
    "# Uncomment if only co-crystal structures / homology models are desired\n",
    "#insider = insider[insider[\"Source\"] != \"ECLAIR\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n",
      "10\n",
      "4754\n",
      "16068\n"
     ]
    }
   ],
   "source": [
    "# Select relevant list of human UniProts\n",
    "# 1. Anything in the SARS2-human interactome\n",
    "# 2. Anything with a disease annotation\n",
    "# 3. Anything with reported human-human interactions in INSIDER\n",
    "keep_unis = set(ires_df[\"P2\"].unique())\n",
    "print len(keep_unis)\n",
    "keep_unis.update(hgmd[\"UNIPROT_ACC\"].unique())\n",
    "print len(keep_unis)\n",
    "keep_unis.update(clinvar[\"UNIPROT_ACC\"].unique())\n",
    "print len(keep_unis)\n",
    "keep_unis.update(insider[\"P1\"].to_list() + insider[\"P2\"].to_list())\n",
    "print len(keep_unis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read in sequence information fo these UniProts\n",
    "##\n",
    "## NOTE: Originally parsed from uniprot_all.fasta, but this takes\n",
    "##       a long time to parse, so I've just saved the relevant sub-set\n",
    "##       and read in from that fasta.\n",
    "##       \n",
    "##       Original uniprot_all.fasta file not provided, but is a combination\n",
    "##       of the sprot and trembl fastas available through the UniProt FTP servers\n",
    "##\n",
    "##       - https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz\n",
    "##       - https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_trembl.fasta.gz\n",
    "#uniall2seq = fasta2dict(\"{0}/uniprot_all.fasta\".format(resource_dir), key_transform=lambda x: x.split(\"|\")[1], keep=keep_unis)\n",
    "#\n",
    "## Save so don't need to parse uniprot_all again\n",
    "#dict2fasta(\"../keep_uniprot.fasta\", uniall2seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in sequence information fo these UniProts\n",
    "uniall2seq = fasta2dict(\"{0}/keep_uniprot.fasta\".format(resource_dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flag any cases where there are discrepencies in the sequences\n",
    "# for these UniProts\n",
    "# (should not happen, but if it does, stick to the seqeunce that\n",
    "#  was used in the rest of this project over the one fetched from\n",
    "#  uniprot_all.fasta)\n",
    "for k in uni2seq:\n",
    "    if(k in uniall2seq):\n",
    "        if(not uni2seq[k] == uniall2seq[k]):\n",
    "            print k\n",
    "            uniall2seq[k] = uni2seq[k]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5d3a1054b09e44fc8ccf2694df674e0a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=122768), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1940ed311e944368afc743144fbf1d8c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=122768), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Populate Ires Human Masks\n",
    "missing = set()\n",
    "wrong_length = set()\n",
    "\n",
    "# Iterate over all interfaces from P1\n",
    "for p1, p1_ires in tqdm_notebook(insider[[\"P1\", \"P1_IRES\"]].values):\n",
    "    try:\n",
    "        # If no mask exists for this protein, initialize an all-zeros mask\n",
    "        if(not p1 in uni2ires_mask_insider):\n",
    "            uni2ires_mask_insider[p1] = np.zeros(len(uniall2seq[p1]))\n",
    "        \n",
    "        \n",
    "        if(not pd.isnull(p1_ires)):\n",
    "            # Parse interface positions\n",
    "            p1_ires = [int(x)-1 for x in unzip_res_range(p1_ires)]\n",
    "            # Automatically drop any interfaces not in range\n",
    "            # NOTE: Decided this is not a safe way to do it. If there are any\n",
    "            #       interfaces out of range, then the whole sequence could be\n",
    "            #       misalinged compared to the sequence used in defining the\n",
    "            #       interface residues in INSIDER. We should just drop these\n",
    "            #       cases to be safe.\n",
    "            #\n",
    "            # NOTE: Does not catch cases where the new sequence is longer than\n",
    "            #       the sequence used from INSIDER\n",
    "            #p1_ires = [int(x)-1 for x in unzip_res_range(p1_ires) if int(x)-1 < len(uniall2seq[p1])]\n",
    "            \n",
    "            # Try to update the mask\n",
    "            try:\n",
    "                uni2ires_mask_insider[p1][p1_ires] = 1\n",
    "            # If any of the interfaces listed in INSIDER are out of range\n",
    "            # based on the length of the current available UniProt sequence then\n",
    "            # we're looking at some odd case (most likely the UniProt sequence\n",
    "            # has changed (different isoform?) between original INSIDER construction\n",
    "            # and now)\n",
    "            except IndexError:\n",
    "                wrong_length.add(p1)\n",
    "                continue\n",
    "    \n",
    "    # If we didn't get a sequence for this protein from our uniprot_all parsing\n",
    "    # just ignore it and move on\n",
    "    except KeyError:\n",
    "        missing.add(p1)\n",
    "\n",
    "# Iterate over all interfaces from P2\n",
    "for p2, p2_ires in tqdm_notebook(insider[[\"P2\", \"P2_IRES\"]].values):\n",
    "    try:\n",
    "        # If no mask exists for this protein, initialize an all-zeros mask\n",
    "        if(not p2 in uni2ires_mask_insider):\n",
    "            uni2ires_mask_insider[p2] = np.zeros(len(uniall2seq[p2]))\n",
    "        \n",
    "        \n",
    "        if(not pd.isnull(p2_ires)):\n",
    "            # Parse interface positions\n",
    "            p2_ires = [int(x)-1 for x in unzip_res_range(p2_ires)]\n",
    "            # Automatically drop any interfaces not in range\n",
    "            # NOTE: Decided this is not a safe way to do it. If there are any\n",
    "            #       interfaces out of range, then the whole sequence could be\n",
    "            #       misalinged compared to the sequence used in defining the\n",
    "            #       interface residues in INSIDER. We should just drop these\n",
    "            #       cases to be safe.\n",
    "            #\n",
    "            # NOTE: Does not catch cases where the new sequence is longer than\n",
    "            #       the sequence used from INSIDER\n",
    "            #p2_ires = [int(x)-1 for x in unzip_res_range(p2_ires) if int(x)-1 < len(uniall2seq[p2])]\n",
    "            \n",
    "            # Try to update the mask\n",
    "            try:\n",
    "                uni2ires_mask_insider[p2][p2_ires] = 1\n",
    "            # If any of the interfaces listed in INSIDER are out of range\n",
    "            # based on the length of the current available UniProt sequence then\n",
    "            # we're looking at some odd case (most likely the UniProt sequence\n",
    "            # has changed (different isoform?) between original INSIDER construction\n",
    "            # and now)\n",
    "            except IndexError:\n",
    "                wrong_length.add(p2)\n",
    "                continue\n",
    "    \n",
    "    # If we didn't get a sequence for this protein from our uniprot_all parsing\n",
    "    # just ignore it and move on\n",
    "    except KeyError:\n",
    "        missing.add(p2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26\n",
      "38\n"
     ]
    }
   ],
   "source": [
    "# Number of UniProts with discrepencies in length / interface annotation\n",
    "print len(wrong_length)\n",
    "\n",
    "# Number of UniProts with sequence unavailable\n",
    "print len(missing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "399a60f7eef14802a1ad7cd65bfeaa77",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=1, bar_style=u'info', max=1), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Populate HGMD Mask\n",
    "c1 = 0 # Number of mutations skipped for data discrepencies\n",
    "c2 = 0 # Number of mutations counted\n",
    "\n",
    "# Iterate over all disease muts\n",
    "for uni, pos, aa_change in tqdm_notebook(hgmd[[\"UNIPROT_ACC\", \"AA_POS\",\"AA_CHANGE\" ]].values):\n",
    "    # If sequence for this Uniprot not available just ignore\n",
    "    if not uni in uniall2seq:\n",
    "        continue\n",
    "    \n",
    "    # If no mask exists for this protein, initialize an all-zeros mask\n",
    "    if(not uni in uni2hgmd_mask):\n",
    "        uni2hgmd_mask[uni] = np.zeros(len(uniall2seq[uni]))\n",
    "    \n",
    "    # If the position of the disease mutant is outside of the range\n",
    "    # expected based on the protein sequence length or there is a\n",
    "    # discrepency in the reference allele at the mutant position between\n",
    "    # HGMD and our UniProt sequence, then skip it (can't trust the\n",
    "    # data)\n",
    "    if(pos > len(uniall2seq[uni]) or uniall2seq[uni][pos-1] != aa_change[0]):\n",
    "        c1 += 1\n",
    "        continue\n",
    "    c2 += 1\n",
    "    uni2hgmd_mask[uni][pos-1] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "0\n",
      "No Mutations\n"
     ]
    }
   ],
   "source": [
    "# Mutations with Discrepencies\n",
    "print c1\n",
    "\n",
    "# Total Mutations\n",
    "print c2\n",
    "\n",
    "# Discrepency Rate\n",
    "try:\n",
    "    print c1 / float(c2)\n",
    "except ZeroDivisionError:\n",
    "    print \"No Mutations\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0e5b3ef9471e4bc9956db826a7376848",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=114675), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Populate ClinVar Mask\n",
    "c1 = 0\n",
    "c2 = 0\n",
    "\n",
    "# Iterate over all disease muts\n",
    "for uni, pos, aa_change in tqdm_notebook(clinvar[[\"UNIPROT_ACC\", \"AA_POS\",\"AA_CHANGE\" ]].values):\n",
    "    # If sequence for this Uniprot not available just ignore\n",
    "    if not uni in uniall2seq:\n",
    "        continue\n",
    "    \n",
    "    # If no mask exists for this protein, initialize an all-zeros mask\n",
    "    if(not uni in uni2clinvar_mask):\n",
    "        uni2clinvar_mask[uni] = np.zeros(len(uniall2seq[uni]))\n",
    "    \n",
    "    # If the position of the disease mutant is outside of the range\n",
    "    # expected based on the protein sequence length or there is a\n",
    "    # discrepency in the reference allele at the mutant position between\n",
    "    # HGMD and our UniProt sequence, then skip it (can't trust the\n",
    "    # data)\n",
    "    if(pos > len(uniall2seq[uni]) or uniall2seq[uni][pos-1] != aa_change[0]):\n",
    "        c1 += 1\n",
    "        continue\n",
    "    c2 += 1\n",
    "    uni2clinvar_mask[uni][pos-1] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1169\n",
      "113202\n",
      "0.0103266726736\n"
     ]
    }
   ],
   "source": [
    "# Mutations with Discrepencies\n",
    "print c1\n",
    "\n",
    "# Total Mutations\n",
    "print c2\n",
    "\n",
    "# Discrepency Rate\n",
    "print c1 / float(c2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "044b250ca83a42a090fd613545d65caf",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=16068), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Calculate Odds Ratios per protein\n",
    "# for enrichment of either ClinVar or HGMD mutations\n",
    "# in either human-viral or human-human interfaces\n",
    "summary = []\n",
    "\n",
    "# Separate masks to store aggregate data\n",
    "# for full proteome vs. only SARS2 interactors\n",
    "insider_hgmd_masks = [[], []]\n",
    "insider_clinvar_masks = [[], []]\n",
    "insider_hgmd_masks_all = [[], []]\n",
    "insider_clinvar_masks_all = [[], []]\n",
    "viral_hgmd_masks = [[], []]\n",
    "viral_clinvar_masks = [[], []]\n",
    "\n",
    "for uni in tqdm_notebook(keep_unis):\n",
    "    if(not uni in uniall2seq):\n",
    "        continue\n",
    "    try:\n",
    "        m_ires_v = uni2ires_mask_viral[uni]\n",
    "    except KeyError:\n",
    "        m_ires_v = np.zeros(len(uniall2seq[uni]))\n",
    "    \n",
    "    try:\n",
    "        m_ires_h = uni2ires_mask_insider[uni]    \n",
    "    except KeyError:\n",
    "        m_ires_h = np.zeros(len(uniall2seq[uni]))\n",
    "    \n",
    "    try:\n",
    "        m_hgmd = uni2hgmd_mask[uni]\n",
    "    except KeyError:\n",
    "        m_hgmd = np.zeros(len(uniall2seq[uni]))\n",
    "    \n",
    "    try:\n",
    "        m_clinvar = uni2clinvar_mask[uni]\n",
    "    except KeyError:\n",
    "        m_clinvar = np.zeros(len(uniall2seq[uni]))\n",
    "        \n",
    "    tmp = [uni, sum(m_ires_v), sum(m_ires_h), sum(m_hgmd), sum(m_clinvar)]\n",
    "    if(tmp[1:] == [0, 0, 0, 0]):\n",
    "        continue\n",
    "    \n",
    "    mask1 = m_ires_v\n",
    "    mask2 = m_hgmd\n",
    "    tmp += odds_ratio(mask1, mask2, log_odds=True)\n",
    "    if(sum(mask1) > 0 and sum(mask2) > 0):\n",
    "        #print uni, \"A\"\n",
    "        viral_hgmd_masks[0].append(mask1)\n",
    "        viral_hgmd_masks[1].append(mask2)\n",
    "    \n",
    "    mask1 = m_ires_v\n",
    "    mask2 = m_clinvar\n",
    "    tmp += odds_ratio(mask1, mask2, log_odds=True)\n",
    "    if(sum(mask1) > 0 and sum(mask2) > 0):\n",
    "        #print uni, \"B\"\n",
    "        viral_clinvar_masks[0].append(mask1)\n",
    "        viral_clinvar_masks[1].append(mask2)\n",
    "    \n",
    "    mask1 = m_ires_h\n",
    "    mask2 = m_hgmd\n",
    "    tmp += odds_ratio(mask1, mask2, log_odds=True)\n",
    "    if(sum(mask1) > 0 and sum(mask2) > 0):\n",
    "        if(uni in uni2seq):\n",
    "            insider_hgmd_masks[0].append(mask1)\n",
    "            insider_hgmd_masks[1].append(mask2)\n",
    "        insider_hgmd_masks_all[0].append(mask1)\n",
    "        insider_hgmd_masks_all[1].append(mask2)\n",
    "\n",
    "    \n",
    "    mask1 = m_ires_h\n",
    "    mask2 = m_clinvar\n",
    "    tmp += odds_ratio(mask1, mask2, log_odds=True)\n",
    "    if(sum(mask1) > 0 and sum(mask2) > 0):\n",
    "        if(uni in uni2seq):\n",
    "            insider_clinvar_masks[0].append(mask1)\n",
    "            insider_clinvar_masks[1].append(mask2)\n",
    "        insider_clinvar_masks_all[0].append(mask1)\n",
    "        insider_clinvar_masks_all[1].append(mask2)\n",
    "\n",
    "    \n",
    "    summary.append(tmp)\n",
    "\n",
    "# We don't really case about the per-protein analysis\n",
    "# on this one, but the setup here is...\n",
    "#\n",
    "# OR_X:Y\n",
    "#\n",
    "# where X can be either H (human-human interfaces from INSIDER)\n",
    "# or V (viral interfaces defined here)\n",
    "#\n",
    "# and Y can be either H (HGMD) or C (ClinVar)\n",
    "summary = pd.DataFrame(summary, columns=[\"UniProt\", \"n_Ires_Viral\", \"n_Ires_Human\", \"n_HGMD\", \"n_Clinvar\", \"OR_V:H\", \"Up_V:H\", \"Low_V:H\", \"p_V:H\", \"OR_V:C\", \"Up_V:C\", \"Low_V:C\", \"p_V:C\", \"OR_H:H\", \"Up_H:H\", \"Low_H:H\", \"p_H:H\", \"OR_H:C\", \"Up_H:C\", \"Low_H:C\", \"p_H:C\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Human Interface (Proteome) ClinVar Muts\n",
      "                Case   Non-Case\n",
      "Exposed      16432.0   412043.0\n",
      "Non-Exposed  71853.0  2557905.0\n",
      "(0.50555557020453712, 0.51826374371010664, 0.49284739669896782, 0.0)\n",
      "\n",
      "Human Interface (SARS2 Interactors) ClinVar Muts\n",
      "             Case  Non-Case\n",
      "Exposed       0.0     335.0\n",
      "Non-Exposed   3.0     447.0\n",
      "(nan, nan, nan, nan)\n",
      "\n",
      "Viral Interface ClinVar Muts\n",
      "             Case  Non-Case\n",
      "Exposed       0.0      54.0\n",
      "Non-Exposed   3.0     728.0\n",
      "(nan, nan, nan, nan)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# View Enrichment within these different categories\n",
    "labels = [\"Human Interface (Proteome) HGMD Muts\", \"Human Interface (SARS2 Interactors) HGMD Muts\", \"Viral Interface HGMD Muts\", \"Human Interface (Proteome) ClinVar Muts\", \"Human Interface (SARS2 Interactors) ClinVar Muts\", \"Viral Interface ClinVar Muts\"]\n",
    "masks = [insider_hgmd_masks_all, insider_hgmd_masks, viral_hgmd_masks, insider_clinvar_masks_all, insider_clinvar_masks, viral_clinvar_masks]\n",
    "for m_set, l in zip(masks, labels):\n",
    "    # Skip HGMD since the data is not provided\n",
    "    if(\"HGMD\" in l):\n",
    "        continue\n",
    "    print l\n",
    "    mask1 = np.concatenate(m_set[0])\n",
    "    mask2 = np.concatenate(m_set[1])\n",
    "    print odds_ratio(mask1, mask2, verbose=True, log_odds=True, error=\"SE\")\n",
    "    print"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt</th>\n",
       "      <th>n_Ires_Viral</th>\n",
       "      <th>n_Ires_Human</th>\n",
       "      <th>n_HGMD</th>\n",
       "      <th>n_Clinvar</th>\n",
       "      <th>OR_V:H</th>\n",
       "      <th>Up_V:H</th>\n",
       "      <th>Low_V:H</th>\n",
       "      <th>p_V:H</th>\n",
       "      <th>OR_V:C</th>\n",
       "      <th>Up_V:C</th>\n",
       "      <th>Low_V:C</th>\n",
       "      <th>p_V:C</th>\n",
       "      <th>OR_H:H</th>\n",
       "      <th>Up_H:H</th>\n",
       "      <th>Low_H:H</th>\n",
       "      <th>p_H:H</th>\n",
       "      <th>OR_H:C</th>\n",
       "      <th>Up_H:C</th>\n",
       "      <th>Low_H:C</th>\n",
       "      <th>p_H:C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [UniProt, n_Ires_Viral, n_Ires_Human, n_HGMD, n_Clinvar, OR_V:H, Up_V:H, Low_V:H, p_V:H, OR_V:C, Up_V:C, Low_V:C, p_V:C, OR_H:H, Up_H:H, Low_H:H, p_H:H, OR_H:C, Up_H:C, Low_H:C, p_H:C]\n",
       "Index: []"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Human Interactors with Enrichment of Disease Mutations\n",
    "# by HGMD in their human-viral Interfaces\n",
    "summary[summary[\"p_V:H\"] <= 0.05].sort_values(\"OR_V:H\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt</th>\n",
       "      <th>n_Ires_Viral</th>\n",
       "      <th>n_Ires_Human</th>\n",
       "      <th>n_HGMD</th>\n",
       "      <th>n_Clinvar</th>\n",
       "      <th>OR_V:H</th>\n",
       "      <th>Up_V:H</th>\n",
       "      <th>Low_V:H</th>\n",
       "      <th>p_V:H</th>\n",
       "      <th>OR_V:C</th>\n",
       "      <th>Up_V:C</th>\n",
       "      <th>Low_V:C</th>\n",
       "      <th>p_V:C</th>\n",
       "      <th>OR_H:H</th>\n",
       "      <th>Up_H:H</th>\n",
       "      <th>Low_H:H</th>\n",
       "      <th>p_H:H</th>\n",
       "      <th>OR_H:C</th>\n",
       "      <th>Up_H:C</th>\n",
       "      <th>Low_H:C</th>\n",
       "      <th>p_H:C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [UniProt, n_Ires_Viral, n_Ires_Human, n_HGMD, n_Clinvar, OR_V:H, Up_V:H, Low_V:H, p_V:H, OR_V:C, Up_V:C, Low_V:C, p_V:C, OR_H:H, Up_H:H, Low_H:H, p_H:H, OR_H:C, Up_H:C, Low_H:C, p_H:C]\n",
       "Index: []"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Human Interactors with Enrichment of Disease Mutations\n",
    "# by ClinVar in their human-viral Interfaces\n",
    "summary[summary[\"p_V:C\"] <= 0.05].sort_values(\"OR_V:C\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt</th>\n",
       "      <th>n_Ires_Viral</th>\n",
       "      <th>n_Ires_Human</th>\n",
       "      <th>n_HGMD</th>\n",
       "      <th>n_Clinvar</th>\n",
       "      <th>OR_V:H</th>\n",
       "      <th>Up_V:H</th>\n",
       "      <th>Low_V:H</th>\n",
       "      <th>p_V:H</th>\n",
       "      <th>OR_V:C</th>\n",
       "      <th>Up_V:C</th>\n",
       "      <th>Low_V:C</th>\n",
       "      <th>p_V:C</th>\n",
       "      <th>OR_H:H</th>\n",
       "      <th>Up_H:H</th>\n",
       "      <th>Low_H:H</th>\n",
       "      <th>p_H:H</th>\n",
       "      <th>OR_H:C</th>\n",
       "      <th>Up_H:C</th>\n",
       "      <th>Low_H:C</th>\n",
       "      <th>p_H:C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [UniProt, n_Ires_Viral, n_Ires_Human, n_HGMD, n_Clinvar, OR_V:H, Up_V:H, Low_V:H, p_V:H, OR_V:C, Up_V:C, Low_V:C, p_V:C, OR_H:H, Up_H:H, Low_H:H, p_H:H, OR_H:C, Up_H:C, Low_H:C, p_H:C]\n",
       "Index: []"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Human Interactors with Enrichment of Disease Mutations\n",
    "# by HGMD in their human-human Interfaces\n",
    "summary[summary[\"p_H:H\"] <= 0.05].sort_values(\"OR_H:H\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt</th>\n",
       "      <th>n_Ires_Viral</th>\n",
       "      <th>n_Ires_Human</th>\n",
       "      <th>n_HGMD</th>\n",
       "      <th>n_Clinvar</th>\n",
       "      <th>OR_V:H</th>\n",
       "      <th>Up_V:H</th>\n",
       "      <th>Low_V:H</th>\n",
       "      <th>p_V:H</th>\n",
       "      <th>OR_V:C</th>\n",
       "      <th>Up_V:C</th>\n",
       "      <th>Low_V:C</th>\n",
       "      <th>p_V:C</th>\n",
       "      <th>OR_H:H</th>\n",
       "      <th>Up_H:H</th>\n",
       "      <th>Low_H:H</th>\n",
       "      <th>p_H:H</th>\n",
       "      <th>OR_H:C</th>\n",
       "      <th>Up_H:C</th>\n",
       "      <th>Low_H:C</th>\n",
       "      <th>p_H:C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1382</th>\n",
       "      <td>P61769</td>\n",
       "      <td>0.0</td>\n",
       "      <td>101.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-3.643856</td>\n",
       "      <td>-0.098026</td>\n",
       "      <td>-7.189687</td>\n",
       "      <td>0.043994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6455</th>\n",
       "      <td>P47897</td>\n",
       "      <td>0.0</td>\n",
       "      <td>311.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.942992</td>\n",
       "      <td>-0.844149</td>\n",
       "      <td>-5.041835</td>\n",
       "      <td>0.005991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14037</th>\n",
       "      <td>P05093</td>\n",
       "      <td>0.0</td>\n",
       "      <td>204.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.668265</td>\n",
       "      <td>-0.544712</td>\n",
       "      <td>-4.791818</td>\n",
       "      <td>0.013789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519</th>\n",
       "      <td>P24752</td>\n",
       "      <td>0.0</td>\n",
       "      <td>162.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.464886</td>\n",
       "      <td>-0.713148</td>\n",
       "      <td>-4.216624</td>\n",
       "      <td>0.005818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11411</th>\n",
       "      <td>Q13496</td>\n",
       "      <td>0.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.353796</td>\n",
       "      <td>-0.862948</td>\n",
       "      <td>-3.844644</td>\n",
       "      <td>0.001972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4420</th>\n",
       "      <td>P40939</td>\n",
       "      <td>0.0</td>\n",
       "      <td>240.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.165465</td>\n",
       "      <td>-0.051696</td>\n",
       "      <td>-4.279235</td>\n",
       "      <td>0.044654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>P29400</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>225.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.157180</td>\n",
       "      <td>-0.482869</td>\n",
       "      <td>-3.831491</td>\n",
       "      <td>0.011563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9743</th>\n",
       "      <td>P16930</td>\n",
       "      <td>0.0</td>\n",
       "      <td>119.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.138561</td>\n",
       "      <td>-0.023028</td>\n",
       "      <td>-4.254094</td>\n",
       "      <td>0.047558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6724</th>\n",
       "      <td>P61244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.882278</td>\n",
       "      <td>-0.197508</td>\n",
       "      <td>-3.567048</td>\n",
       "      <td>0.028544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1457</th>\n",
       "      <td>Q9UMX1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>162.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.854097</td>\n",
       "      <td>-0.313737</td>\n",
       "      <td>-3.394458</td>\n",
       "      <td>0.018316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2357</th>\n",
       "      <td>Q09472</td>\n",
       "      <td>0.0</td>\n",
       "      <td>681.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.845505</td>\n",
       "      <td>-0.783083</td>\n",
       "      <td>-2.907927</td>\n",
       "      <td>0.000663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1819</th>\n",
       "      <td>Q01968</td>\n",
       "      <td>0.0</td>\n",
       "      <td>210.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.583489</td>\n",
       "      <td>-0.080047</td>\n",
       "      <td>-3.086931</td>\n",
       "      <td>0.038988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12160</th>\n",
       "      <td>P00441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>101.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.417955</td>\n",
       "      <td>-0.214071</td>\n",
       "      <td>-2.621839</td>\n",
       "      <td>0.020973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8402</th>\n",
       "      <td>Q03426</td>\n",
       "      <td>0.0</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.338754</td>\n",
       "      <td>-0.264114</td>\n",
       "      <td>-2.413393</td>\n",
       "      <td>0.014620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12202</th>\n",
       "      <td>P04637</td>\n",
       "      <td>0.0</td>\n",
       "      <td>280.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>229.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.174791</td>\n",
       "      <td>-0.494229</td>\n",
       "      <td>-1.855353</td>\n",
       "      <td>0.000716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8222</th>\n",
       "      <td>Q08289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>270.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.112266</td>\n",
       "      <td>-0.235219</td>\n",
       "      <td>-1.989313</td>\n",
       "      <td>0.012933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14952</th>\n",
       "      <td>P41250</td>\n",
       "      <td>0.0</td>\n",
       "      <td>328.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.080920</td>\n",
       "      <td>-0.130204</td>\n",
       "      <td>-2.031636</td>\n",
       "      <td>0.025855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6383</th>\n",
       "      <td>P35557</td>\n",
       "      <td>0.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.055761</td>\n",
       "      <td>-0.236469</td>\n",
       "      <td>-1.875053</td>\n",
       "      <td>0.011548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9590</th>\n",
       "      <td>P15289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.038848</td>\n",
       "      <td>-0.185175</td>\n",
       "      <td>-1.892521</td>\n",
       "      <td>0.017074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13412</th>\n",
       "      <td>P19971</td>\n",
       "      <td>0.0</td>\n",
       "      <td>148.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.996609</td>\n",
       "      <td>-0.041023</td>\n",
       "      <td>-1.952195</td>\n",
       "      <td>0.040943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4283</th>\n",
       "      <td>P09493</td>\n",
       "      <td>0.0</td>\n",
       "      <td>179.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.936352</td>\n",
       "      <td>-0.203231</td>\n",
       "      <td>-1.669473</td>\n",
       "      <td>0.012304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9064</th>\n",
       "      <td>P13637</td>\n",
       "      <td>0.0</td>\n",
       "      <td>427.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.933474</td>\n",
       "      <td>-0.102516</td>\n",
       "      <td>-1.764431</td>\n",
       "      <td>0.027682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11387</th>\n",
       "      <td>P11310</td>\n",
       "      <td>0.0</td>\n",
       "      <td>172.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.808817</td>\n",
       "      <td>-0.038789</td>\n",
       "      <td>-1.578845</td>\n",
       "      <td>0.039523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8884</th>\n",
       "      <td>P02461</td>\n",
       "      <td>0.0</td>\n",
       "      <td>240.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.583410</td>\n",
       "      <td>-0.041021</td>\n",
       "      <td>-1.125798</td>\n",
       "      <td>0.035014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6293</th>\n",
       "      <td>P35555</td>\n",
       "      <td>0.0</td>\n",
       "      <td>414.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>536.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.471829</td>\n",
       "      <td>-0.053989</td>\n",
       "      <td>-0.889670</td>\n",
       "      <td>0.026883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13943</th>\n",
       "      <td>P21359</td>\n",
       "      <td>0.0</td>\n",
       "      <td>217.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>627.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.469988</td>\n",
       "      <td>0.919629</td>\n",
       "      <td>0.020347</td>\n",
       "      <td>0.040496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4741</th>\n",
       "      <td>P11532</td>\n",
       "      <td>0.0</td>\n",
       "      <td>346.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>371.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.507911</td>\n",
       "      <td>0.985112</td>\n",
       "      <td>0.030710</td>\n",
       "      <td>0.036970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7362</th>\n",
       "      <td>P04275</td>\n",
       "      <td>0.0</td>\n",
       "      <td>404.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.527982</td>\n",
       "      <td>1.009100</td>\n",
       "      <td>0.046864</td>\n",
       "      <td>0.031486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14711</th>\n",
       "      <td>Q8WZ42</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1118.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3117.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.528486</td>\n",
       "      <td>0.790474</td>\n",
       "      <td>0.266498</td>\n",
       "      <td>0.000077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1247</th>\n",
       "      <td>P35609</td>\n",
       "      <td>0.0</td>\n",
       "      <td>396.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>136.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.541307</td>\n",
       "      <td>1.069215</td>\n",
       "      <td>0.013398</td>\n",
       "      <td>0.044462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10164</th>\n",
       "      <td>Q9NR61</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.042207</td>\n",
       "      <td>6.573713</td>\n",
       "      <td>1.510700</td>\n",
       "      <td>0.001751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6364</th>\n",
       "      <td>P23508</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.075033</td>\n",
       "      <td>7.417227</td>\n",
       "      <td>0.732840</td>\n",
       "      <td>0.016861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8466</th>\n",
       "      <td>Q9ULG1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.099945</td>\n",
       "      <td>7.589968</td>\n",
       "      <td>0.609923</td>\n",
       "      <td>0.021307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6652</th>\n",
       "      <td>Q15392</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.178642</td>\n",
       "      <td>6.696531</td>\n",
       "      <td>1.660754</td>\n",
       "      <td>0.001143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7498</th>\n",
       "      <td>Q9H3S5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.216525</td>\n",
       "      <td>7.785790</td>\n",
       "      <td>0.647261</td>\n",
       "      <td>0.020592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10013</th>\n",
       "      <td>P05546</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.222392</td>\n",
       "      <td>7.712043</td>\n",
       "      <td>0.732742</td>\n",
       "      <td>0.017716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6650</th>\n",
       "      <td>O14841</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.252665</td>\n",
       "      <td>7.103451</td>\n",
       "      <td>1.401880</td>\n",
       "      <td>0.003458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10590</th>\n",
       "      <td>Q8WVX9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.262456</td>\n",
       "      <td>7.815525</td>\n",
       "      <td>0.709387</td>\n",
       "      <td>0.018709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11927</th>\n",
       "      <td>Q9H4F8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.283640</td>\n",
       "      <td>7.181187</td>\n",
       "      <td>1.386092</td>\n",
       "      <td>0.003761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10005</th>\n",
       "      <td>P05549</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.354029</td>\n",
       "      <td>6.455712</td>\n",
       "      <td>2.252346</td>\n",
       "      <td>0.000049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2857</th>\n",
       "      <td>Q9BWT7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.373590</td>\n",
       "      <td>7.004333</td>\n",
       "      <td>1.742848</td>\n",
       "      <td>0.001120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5435</th>\n",
       "      <td>Q8NFF5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.426265</td>\n",
       "      <td>8.465177</td>\n",
       "      <td>0.387353</td>\n",
       "      <td>0.031719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13936</th>\n",
       "      <td>Q8WXG6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.428561</td>\n",
       "      <td>8.441739</td>\n",
       "      <td>0.415383</td>\n",
       "      <td>0.030554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14651</th>\n",
       "      <td>Q9H0W5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.445705</td>\n",
       "      <td>8.005876</td>\n",
       "      <td>0.885533</td>\n",
       "      <td>0.014386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12361</th>\n",
       "      <td>P35716</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.473237</td>\n",
       "      <td>7.764922</td>\n",
       "      <td>1.181553</td>\n",
       "      <td>0.007733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2461</th>\n",
       "      <td>Q99592</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.564785</td>\n",
       "      <td>7.461734</td>\n",
       "      <td>1.667835</td>\n",
       "      <td>0.002013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4012</th>\n",
       "      <td>Q719H9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.584963</td>\n",
       "      <td>7.671702</td>\n",
       "      <td>1.498223</td>\n",
       "      <td>0.003600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2330</th>\n",
       "      <td>P56645</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.740241</td>\n",
       "      <td>8.256413</td>\n",
       "      <td>1.224068</td>\n",
       "      <td>0.008235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6093</th>\n",
       "      <td>Q9UNQ0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.775646</td>\n",
       "      <td>8.819321</td>\n",
       "      <td>0.731971</td>\n",
       "      <td>0.020627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12118</th>\n",
       "      <td>Q9UMN6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.950359</td>\n",
       "      <td>8.961355</td>\n",
       "      <td>0.939362</td>\n",
       "      <td>0.015564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9543</th>\n",
       "      <td>Q5K651</td>\n",
       "      <td>0.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.098734</td>\n",
       "      <td>9.120406</td>\n",
       "      <td>1.077062</td>\n",
       "      <td>0.012960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9532</th>\n",
       "      <td>Q70Z35</td>\n",
       "      <td>0.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.216861</td>\n",
       "      <td>9.240103</td>\n",
       "      <td>1.193620</td>\n",
       "      <td>0.011039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10554</th>\n",
       "      <td>Q8TEA7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.237989</td>\n",
       "      <td>9.281234</td>\n",
       "      <td>1.194745</td>\n",
       "      <td>0.011113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1407</th>\n",
       "      <td>Q9UJV3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.451211</td>\n",
       "      <td>8.951857</td>\n",
       "      <td>1.950565</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2103</th>\n",
       "      <td>Q14202</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.528959</td>\n",
       "      <td>9.562898</td>\n",
       "      <td>1.495020</td>\n",
       "      <td>0.007224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9540</th>\n",
       "      <td>Q96RN1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.580447</td>\n",
       "      <td>9.158316</td>\n",
       "      <td>2.002578</td>\n",
       "      <td>0.002236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15170</th>\n",
       "      <td>Q7Z3V4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.756293</td>\n",
       "      <td>9.250779</td>\n",
       "      <td>2.261807</td>\n",
       "      <td>0.001244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11442</th>\n",
       "      <td>P48764</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.879146</td>\n",
       "      <td>9.504771</td>\n",
       "      <td>2.253521</td>\n",
       "      <td>0.001482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8311</th>\n",
       "      <td>Q9BWT1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.185867</td>\n",
       "      <td>10.382538</td>\n",
       "      <td>1.989195</td>\n",
       "      <td>0.003865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684</th>\n",
       "      <td>P28335</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.820179</td>\n",
       "      <td>11.063709</td>\n",
       "      <td>2.576649</td>\n",
       "      <td>0.001632</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>295 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      UniProt  n_Ires_Viral  n_Ires_Human  n_HGMD  n_Clinvar  OR_V:H  Up_V:H  \\\n",
       "1382   P61769           0.0         101.0     0.0        3.0     NaN     NaN   \n",
       "6455   P47897           0.0         311.0     0.0       24.0     NaN     NaN   \n",
       "14037  P05093           0.0         204.0     0.0       20.0     NaN     NaN   \n",
       "3519   P24752           0.0         162.0     0.0       28.0     NaN     NaN   \n",
       "11411  Q13496           0.0         156.0     0.0       57.0     NaN     NaN   \n",
       "4420   P40939           0.0         240.0     0.0       21.0     NaN     NaN   \n",
       "86     P29400           0.0          86.0     0.0      225.0     NaN     NaN   \n",
       "9743   P16930           0.0         119.0     0.0       23.0     NaN     NaN   \n",
       "6724   P61244           0.0          80.0     0.0       17.0     NaN     NaN   \n",
       "1457   Q9UMX1           0.0         162.0     0.0       31.0     NaN     NaN   \n",
       "2357   Q09472           0.0         681.0     0.0       79.0     NaN     NaN   \n",
       "1819   Q01968           0.0         210.0     0.0       42.0     NaN     NaN   \n",
       "12160  P00441           0.0         101.0     0.0       28.0     NaN     NaN   \n",
       "8402   Q03426           0.0         111.0     0.0       61.0     NaN     NaN   \n",
       "12202  P04637           0.0         280.0     0.0      229.0     NaN     NaN   \n",
       "8222   Q08289           0.0         270.0     0.0       59.0     NaN     NaN   \n",
       "14952  P41250           0.0         328.0     0.0       46.0     NaN     NaN   \n",
       "6383   P35557           0.0         126.0     0.0      100.0     NaN     NaN   \n",
       "9590   P15289           0.0         127.0     0.0       97.0     NaN     NaN   \n",
       "13412  P19971           0.0         148.0     0.0       62.0     NaN     NaN   \n",
       "4283   P09493           0.0         179.0     0.0       93.0     NaN     NaN   \n",
       "9064   P13637           0.0         427.0     0.0       60.0     NaN     NaN   \n",
       "11387  P11310           0.0         172.0     0.0       76.0     NaN     NaN   \n",
       "8884   P02461           0.0         240.0     0.0      300.0     NaN     NaN   \n",
       "6293   P35555           0.0         414.0     0.0      536.0     NaN     NaN   \n",
       "13943  P21359           0.0         217.0     0.0      627.0     NaN     NaN   \n",
       "4741   P11532           0.0         346.0     0.0      371.0     NaN     NaN   \n",
       "7362   P04275           0.0         404.0     0.0      254.0     NaN     NaN   \n",
       "14711  Q8WZ42           0.0        1118.0     0.0     3117.0     NaN     NaN   \n",
       "1247   P35609           0.0         396.0     0.0      136.0     NaN     NaN   \n",
       "...       ...           ...           ...     ...        ...     ...     ...   \n",
       "10164  Q9NR61           0.0          22.0     0.0        6.0     NaN     NaN   \n",
       "6364   P23508           0.0          17.0     0.0        4.0     NaN     NaN   \n",
       "8466   Q9ULG1           0.0          45.0     0.0        3.0     NaN     NaN   \n",
       "6652   Q15392           0.0           7.0     0.0       13.0     NaN     NaN   \n",
       "7498   Q9H3S5           0.0          12.0     0.0        3.0     NaN     NaN   \n",
       "10013  P05546           0.0          50.0     0.0        3.0     NaN     NaN   \n",
       "6650   O14841           0.0          66.0     0.0        4.0     NaN     NaN   \n",
       "10590  Q8WVX9           0.0          14.0     0.0        3.0     NaN     NaN   \n",
       "11927  Q9H4F8           0.0          23.0     0.0        4.0     NaN     NaN   \n",
       "10005  P05549           0.0          24.0     0.0        8.0     NaN     NaN   \n",
       "2857   Q9BWT7           0.0          34.0     0.0        5.0     NaN     NaN   \n",
       "5435   Q8NFF5           0.0          27.0     0.0        2.0     NaN     NaN   \n",
       "13936  Q8WXG6           0.0          74.0     0.0        2.0     NaN     NaN   \n",
       "14651  Q9H0W5           0.0          13.0     0.0        3.0     NaN     NaN   \n",
       "12361  P35716           0.0          55.0     0.0        4.0     NaN     NaN   \n",
       "2461   Q99592           0.0          23.0     0.0        4.0     NaN     NaN   \n",
       "4012   Q719H9           0.0          56.0     0.0        7.0     NaN     NaN   \n",
       "2330   P56645           0.0          23.0     0.0        3.0     NaN     NaN   \n",
       "6093   Q9UNQ0           0.0          24.0     0.0        2.0     NaN     NaN   \n",
       "12118  Q9UMN6           0.0          86.0     0.0        2.0     NaN     NaN   \n",
       "9543   Q5K651           0.0          46.0     0.0        2.0     NaN     NaN   \n",
       "9532   Q70Z35           0.0          43.0     0.0        2.0     NaN     NaN   \n",
       "10554  Q8TEA7           0.0          24.0     0.0        2.0     NaN     NaN   \n",
       "1407   Q9UJV3           0.0          34.0     0.0        3.0     NaN     NaN   \n",
       "2103   Q14202           0.0          30.0     0.0        2.0     NaN     NaN   \n",
       "9540   Q96RN1           0.0          11.0     0.0        3.0     NaN     NaN   \n",
       "15170  Q7Z3V4           0.0          40.0     0.0        3.0     NaN     NaN   \n",
       "11442  P48764           0.0           8.0     0.0        3.0     NaN     NaN   \n",
       "8311   Q9BWT1           0.0           6.0     0.0        2.0     NaN     NaN   \n",
       "684    P28335           0.0           5.0     0.0        2.0     NaN     NaN   \n",
       "\n",
       "       Low_V:H  p_V:H  OR_V:C  Up_V:C  Low_V:C  p_V:C  OR_H:H  Up_H:H  \\\n",
       "1382       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "6455       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "14037      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "3519       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "11411      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "4420       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "86         NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "9743       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "6724       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "1457       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "2357       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "1819       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "12160      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "8402       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "12202      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "8222       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "14952      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "6383       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "9590       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "13412      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "4283       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "9064       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "11387      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "8884       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "6293       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "13943      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "4741       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "7362       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "14711      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "1247       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "...        ...    ...     ...     ...      ...    ...     ...     ...   \n",
       "10164      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "6364       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "8466       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "6652       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "7498       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "10013      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "6650       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "10590      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "11927      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "10005      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "2857       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "5435       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "13936      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "14651      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "12361      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "2461       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "4012       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "2330       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "6093       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "12118      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "9543       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "9532       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "10554      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "1407       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "2103       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "9540       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "15170      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "11442      NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "8311       NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "684        NaN    NaN     NaN     NaN      NaN    NaN     NaN     NaN   \n",
       "\n",
       "       Low_H:H  p_H:H    OR_H:C     Up_H:C   Low_H:C     p_H:C  \n",
       "1382       NaN    NaN -3.643856  -0.098026 -7.189687  0.043994  \n",
       "6455       NaN    NaN -2.942992  -0.844149 -5.041835  0.005991  \n",
       "14037      NaN    NaN -2.668265  -0.544712 -4.791818  0.013789  \n",
       "3519       NaN    NaN -2.464886  -0.713148 -4.216624  0.005818  \n",
       "11411      NaN    NaN -2.353796  -0.862948 -3.844644  0.001972  \n",
       "4420       NaN    NaN -2.165465  -0.051696 -4.279235  0.044654  \n",
       "86         NaN    NaN -2.157180  -0.482869 -3.831491  0.011563  \n",
       "9743       NaN    NaN -2.138561  -0.023028 -4.254094  0.047558  \n",
       "6724       NaN    NaN -1.882278  -0.197508 -3.567048  0.028544  \n",
       "1457       NaN    NaN -1.854097  -0.313737 -3.394458  0.018316  \n",
       "2357       NaN    NaN -1.845505  -0.783083 -2.907927  0.000663  \n",
       "1819       NaN    NaN -1.583489  -0.080047 -3.086931  0.038988  \n",
       "12160      NaN    NaN -1.417955  -0.214071 -2.621839  0.020973  \n",
       "8402       NaN    NaN -1.338754  -0.264114 -2.413393  0.014620  \n",
       "12202      NaN    NaN -1.174791  -0.494229 -1.855353  0.000716  \n",
       "8222       NaN    NaN -1.112266  -0.235219 -1.989313  0.012933  \n",
       "14952      NaN    NaN -1.080920  -0.130204 -2.031636  0.025855  \n",
       "6383       NaN    NaN -1.055761  -0.236469 -1.875053  0.011548  \n",
       "9590       NaN    NaN -1.038848  -0.185175 -1.892521  0.017074  \n",
       "13412      NaN    NaN -0.996609  -0.041023 -1.952195  0.040943  \n",
       "4283       NaN    NaN -0.936352  -0.203231 -1.669473  0.012304  \n",
       "9064       NaN    NaN -0.933474  -0.102516 -1.764431  0.027682  \n",
       "11387      NaN    NaN -0.808817  -0.038789 -1.578845  0.039523  \n",
       "8884       NaN    NaN -0.583410  -0.041021 -1.125798  0.035014  \n",
       "6293       NaN    NaN -0.471829  -0.053989 -0.889670  0.026883  \n",
       "13943      NaN    NaN  0.469988   0.919629  0.020347  0.040496  \n",
       "4741       NaN    NaN  0.507911   0.985112  0.030710  0.036970  \n",
       "7362       NaN    NaN  0.527982   1.009100  0.046864  0.031486  \n",
       "14711      NaN    NaN  0.528486   0.790474  0.266498  0.000077  \n",
       "1247       NaN    NaN  0.541307   1.069215  0.013398  0.044462  \n",
       "...        ...    ...       ...        ...       ...       ...  \n",
       "10164      NaN    NaN  4.042207   6.573713  1.510700  0.001751  \n",
       "6364       NaN    NaN  4.075033   7.417227  0.732840  0.016861  \n",
       "8466       NaN    NaN  4.099945   7.589968  0.609923  0.021307  \n",
       "6652       NaN    NaN  4.178642   6.696531  1.660754  0.001143  \n",
       "7498       NaN    NaN  4.216525   7.785790  0.647261  0.020592  \n",
       "10013      NaN    NaN  4.222392   7.712043  0.732742  0.017716  \n",
       "6650       NaN    NaN  4.252665   7.103451  1.401880  0.003458  \n",
       "10590      NaN    NaN  4.262456   7.815525  0.709387  0.018709  \n",
       "11927      NaN    NaN  4.283640   7.181187  1.386092  0.003761  \n",
       "10005      NaN    NaN  4.354029   6.455712  2.252346  0.000049  \n",
       "2857       NaN    NaN  4.373590   7.004333  1.742848  0.001120  \n",
       "5435       NaN    NaN  4.426265   8.465177  0.387353  0.031719  \n",
       "13936      NaN    NaN  4.428561   8.441739  0.415383  0.030554  \n",
       "14651      NaN    NaN  4.445705   8.005876  0.885533  0.014386  \n",
       "12361      NaN    NaN  4.473237   7.764922  1.181553  0.007733  \n",
       "2461       NaN    NaN  4.564785   7.461734  1.667835  0.002013  \n",
       "4012       NaN    NaN  4.584963   7.671702  1.498223  0.003600  \n",
       "2330       NaN    NaN  4.740241   8.256413  1.224068  0.008235  \n",
       "6093       NaN    NaN  4.775646   8.819321  0.731971  0.020627  \n",
       "12118      NaN    NaN  4.950359   8.961355  0.939362  0.015564  \n",
       "9543       NaN    NaN  5.098734   9.120406  1.077062  0.012960  \n",
       "9532       NaN    NaN  5.216861   9.240103  1.193620  0.011039  \n",
       "10554      NaN    NaN  5.237989   9.281234  1.194745  0.011113  \n",
       "1407       NaN    NaN  5.451211   8.951857  1.950565  0.002273  \n",
       "2103       NaN    NaN  5.528959   9.562898  1.495020  0.007224  \n",
       "9540       NaN    NaN  5.580447   9.158316  2.002578  0.002236  \n",
       "15170      NaN    NaN  5.756293   9.250779  2.261807  0.001244  \n",
       "11442      NaN    NaN  5.879146   9.504771  2.253521  0.001482  \n",
       "8311       NaN    NaN  6.185867  10.382538  1.989195  0.003865  \n",
       "684        NaN    NaN  6.820179  11.063709  2.576649  0.001632  \n",
       "\n",
       "[295 rows x 21 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Human Interactors with Enrichment of Disease Mutations\n",
    "# by ClinVar in their human-human Interfaces\n",
    "summary[summary[\"p_H:C\"] <= 0.05].sort_values(\"OR_H:C\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyze Term Enrichment Based on Disease Ontology DAG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The goal here is to re-do the enrichment calculations looking at disease terms\n",
    "# in the context of a full disease ontology. Where for instance, a particular variant\n",
    "# may only be listed under a specific phenotype \"Colorectal Cancer\" but could also\n",
    "# apply to a more general \"Cancer\" term.\n",
    "#\n",
    "# In this way, when we count a term, we also add that count to all its parents so that\n",
    "# we can accurately look at enrichment at all levels of specificity. This avoids running\n",
    "# into issues where we have low counts (can't do good statistical analysis) from many specific\n",
    "# terms, and could instead do a good enrichment test on a more general common parent term."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CUI1</th>\n",
       "      <th>AUI1</th>\n",
       "      <th>STYPE1</th>\n",
       "      <th>REL</th>\n",
       "      <th>CUI2</th>\n",
       "      <th>AUI2</th>\n",
       "      <th>RELA</th>\n",
       "      <th>RUI</th>\n",
       "      <th>SAB</th>\n",
       "      <th>SL</th>\n",
       "      <th>SUPPRESS</th>\n",
       "      <th>Unnamed: 11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C0000727</td>\n",
       "      <td>A0017734</td>\n",
       "      <td>AUI</td>\n",
       "      <td>SY</td>\n",
       "      <td>C0000727</td>\n",
       "      <td>A0017736</td>\n",
       "      <td>permuted_term_of</td>\n",
       "      <td>R28483132</td>\n",
       "      <td>MSH</td>\n",
       "      <td>MSH</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C0000727</td>\n",
       "      <td>A0017734</td>\n",
       "      <td>AUI</td>\n",
       "      <td>SY</td>\n",
       "      <td>C0000727</td>\n",
       "      <td>A0019566</td>\n",
       "      <td>permuted_term_of</td>\n",
       "      <td>R28483133</td>\n",
       "      <td>MSH</td>\n",
       "      <td>MSH</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C0000727</td>\n",
       "      <td>A0017734</td>\n",
       "      <td>AUI</td>\n",
       "      <td>SY</td>\n",
       "      <td>C0000727</td>\n",
       "      <td>A0019567</td>\n",
       "      <td>permuted_term_of</td>\n",
       "      <td>R28483134</td>\n",
       "      <td>MSH</td>\n",
       "      <td>MSH</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C0000727</td>\n",
       "      <td>A0017734</td>\n",
       "      <td>SDUI</td>\n",
       "      <td>PAR</td>\n",
       "      <td>C0000737</td>\n",
       "      <td>A0017753</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R71382266</td>\n",
       "      <td>MSH</td>\n",
       "      <td>MSH</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C0000727</td>\n",
       "      <td>A0017734</td>\n",
       "      <td>SDUI</td>\n",
       "      <td>AQ</td>\n",
       "      <td>C0007994</td>\n",
       "      <td>A3879713</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R120176394</td>\n",
       "      <td>MSH</td>\n",
       "      <td>MSH</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      #CUI1      AUI1 STYPE1  REL      CUI2      AUI2              RELA  \\\n",
       "0  C0000727  A0017734    AUI   SY  C0000727  A0017736  permuted_term_of   \n",
       "1  C0000727  A0017734    AUI   SY  C0000727  A0019566  permuted_term_of   \n",
       "2  C0000727  A0017734    AUI   SY  C0000727  A0019567  permuted_term_of   \n",
       "3  C0000727  A0017734   SDUI  PAR  C0000737  A0017753               NaN   \n",
       "4  C0000727  A0017734   SDUI   AQ  C0007994  A3879713               NaN   \n",
       "\n",
       "          RUI  SAB   SL SUPPRESS  Unnamed: 11  \n",
       "0   R28483132  MSH  MSH        N          NaN  \n",
       "1   R28483133  MSH  MSH        N          NaN  \n",
       "2   R28483134  MSH  MSH        N          NaN  \n",
       "3   R71382266  MSH  MSH        N          NaN  \n",
       "4  R120176394  MSH  MSH        N          NaN  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Download and Read in Raw MedGen Relationships (these files too large to include in repository)\n",
    "if(not os.path.exists(\"{0}/MGREL.RRF\".format(resource_dir))):\n",
    "    os.system(\"wget https://ftp.ncbi.nlm.nih.gov/pub/medgen/MGREL.RRF.gz --directory-prefix {0}\".format(resource_dir))\n",
    "    os.system(\"gunzip {0}/MGREL.RRF.gz\".format(resource_dir))\n",
    "medgen_rel = pd.read_csv(\"{0}/MGREL.RRF\".format(resource_dir), sep=\"|\")\n",
    "medgen_rel.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "102120\n",
      "62838\n"
     ]
    }
   ],
   "source": [
    "# Total Number of Terms\n",
    "all_terms = set(medgen_rel[\"#CUI1\"].to_list() + medgen_rel[\"CUI2\"].to_list())\n",
    "print len(all_terms)\n",
    "\n",
    "# Only Terms involved in Parent / Child Relationships\n",
    "child_parent_terms = set(medgen_rel[medgen_rel[\"REL\"].isin([\"PAR\", \"CHD\"])][\"#CUI1\"].to_list() + medgen_rel[medgen_rel[\"REL\"].isin([\"PAR\", \"CHD\"])][\"CUI2\"].to_list())\n",
    "print len(child_parent_terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6275\n",
      "2945\n",
      "4712\n"
     ]
    }
   ],
   "source": [
    "# NOTE: Only the ClinVar dataset lists MedGen terms, so we\n",
    "#       only work with this dataset here\n",
    "\n",
    "# Subset of MedGen Terms from ClinVar\n",
    "clinvar_terms = set(flatten(clinvar[\"PHEN_IDS\"].map(lambda x: [y.split(\":\")[1] for y in str(x).replace(\";\", \",\").split(\",\") if y.split(\":\")[0] == \"MedGen\"]).to_list()))\n",
    "print len(clinvar_terms)\n",
    "\n",
    "# Overlap with Parent / Child Terms\n",
    "print len(clinvar_terms.intersection(child_parent_terms))\n",
    "\n",
    "# Overlap with All Terms from MedGen Relationships\n",
    "# I'm unsure what's up with the ~1500 that are missing\n",
    "# from the MedGen REL file (maybe they just aren't in any\n",
    "# relationships?)\n",
    "print len(clinvar_terms.intersection(all_terms))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c166913983d545a993c835dbaa161586",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=1117150), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1dd9ed52fb554b4b96b171759e61a7b0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=222872), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Here we being constructing a directed graph for the disease ontology\n",
    "# We first add in all the parent child relationships listed in the MedGene\n",
    "# relationships.\n",
    "#\n",
    "dag = nx.DiGraph()\n",
    "\n",
    "# Add all the Nodes.\n",
    "# Each node stores the counts that will be used for its enrichment\n",
    "# We start these as out with no disease gene counts (e.g. everything\n",
    "# is either an interacotr non-disease gene (b) or a non-interactor\n",
    "# non-disease gene (d)). As we parse the graph we will update these\n",
    "# counts by adding / subtracting as necessary and propogating the\n",
    "# changed through to the parent nodes\n",
    "for c1, c2 in tqdm_notebook(medgen_rel[[\"#CUI1\", \"CUI2\"]].values):\n",
    "    dag.add_node(c1, **{\"a\":0, \"b\":n_interactors, \"c\":0, \"d\":n_proteome-n_interactors, \"block\":False})\n",
    "    dag.add_node(c2, **{\"a\":0, \"b\":n_interactors, \"c\":0, \"d\":n_proteome-n_interactors, \"block\":False})\n",
    "\n",
    "# Add all the relationships\n",
    "for c1, c2, rel in tqdm_notebook(medgen_rel[medgen_rel[\"REL\"].isin([\"PAR\", \"CHD\"])][[\"#CUI1\", \"CUI2\", \"REL\"]].values):\n",
    "    if(rel == \"PAR\"):\n",
    "        dag.add_edge(c2, c1)\n",
    "    else:\n",
    "        dag.add_edge(c1, c2)\n",
    "\n",
    "# Add in any extra terms from ClinVar Table with no relationships\n",
    "for c in set(flatten(clinvar[\"PHEN_IDS\"].map(lambda x: [y.split(\":\")[1] for y in str(x).replace(\";\", \",\").split(\",\") if y.split(\":\")[0] == \"MedGen\"]).to_list())):\n",
    "    dag.add_node(c, **{\"a\":0, \"b\":n_interactors, \"c\":0, \"d\":n_proteome-n_interactors, \"block\":False})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Updates the counts for a node and its predecessors\n",
    "def propogate_updates(g, n, is_interactor):\n",
    "    # If the node doesn't exist, initialize it\n",
    "    if(not n in g.nodes):\n",
    "        g.add_node(n, **{\"a\":0, \"b\":322, \"c\":0, \"d\":20350-332, \"block\":False})\n",
    "    #print n\n",
    "    \n",
    "    # If the node has been blocked (i.e. has already been visited\n",
    "    # in this round of updates, then stop)\n",
    "    if(g.node[n][\"block\"]):\n",
    "        #print \"Already Reached\", n\n",
    "        return\n",
    "    \n",
    "    # Otherwise, update the count as necessary depending on if this gene\n",
    "    # is / is not an interactor of SARS2\n",
    "    else:\n",
    "        if(is_interactor):\n",
    "            g.node[n][\"a\"] += 1\n",
    "            g.node[n][\"b\"] -= 1\n",
    "        else:\n",
    "            g.node[n][\"c\"] += 1\n",
    "            g.node[n][\"d\"] -= 1\n",
    "        \n",
    "        # Block this node so that it is not revisited\n",
    "        g.node[n][\"block\"] = True\n",
    "        \n",
    "        # Propogate the update through all parent terms\n",
    "        for p in g.predecessors(n):\n",
    "            propogate_updates(g, p, is_interactor)\n",
    "# FUNCTION END\n",
    "\n",
    "# Unblocks all the nodes in the network\n",
    "def unblock_all(g):\n",
    "    for n in g.nodes:\n",
    "        g.node[n][\"block\"] = False\n",
    "# FUNCTION END\n",
    "\n",
    "# Resets all the nodes in the network to their initial conditions\n",
    "def reset_all(g):\n",
    "    for n in g.nodes:\n",
    "        for k, v in {\"a\":0, \"b\":n_interactors, \"c\":0, \"d\":n_proteome-n_interactors, \"block\":False}.iteritems():\n",
    "            g.node[n][k] = v\n",
    "# FUNCTION END"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make sure the network is reinitialized with starting counts\n",
    "reset_all(dag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7481fd377d9b4cd68efda3c3e3dd70fb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=4747), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Iterate through all Uniprot IDs in the ClinVar Data\n",
    "for uni, g in tqdm_notebook(clinvar[[\"UNIPROT_ACC\", \"PHEN_IDS\", \"PHEN\"]].groupby(\"UNIPROT_ACC\")):\n",
    "    # Fetch all the MedGen terms associated with this gene\n",
    "    med_gen_terms = set(flatten(g[\"PHEN_IDS\"].map(lambda x: [y.split(\":\")[1] for y in str(x).replace(\";\", \",\").split(\",\") if y.split(\":\")[0] == \"MedGen\"]).to_list()))\n",
    "    \n",
    "    # Figure out if this is an interactor\n",
    "    is_interactor = uni in uni2seq.keys()\n",
    "    \n",
    "    # Udate the gene counts for all terms\n",
    "    # this gene is invovled in and its parents\n",
    "    for n in med_gen_terms:\n",
    "        propogate_updates(dag, n, is_interactor)\n",
    "    \n",
    "    # Unblock all nodes in the network to prepare for adding the next gene\n",
    "    unblock_all(dag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3122408b9bb24a439e01834a652f0fd6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=103683), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Store initial summary reporting each term, the counts, and its\n",
    "# list of immediate children and parents.\n",
    "summary = []\n",
    "for n in tqdm_notebook(dag.nodes):\n",
    "    summary.append([n, dag.nodes[n][\"a\"], dag.nodes[n][\"b\"], dag.nodes[n][\"c\"], dag.nodes[n][\"d\"], \",\".join(dag.successors(n)), \",\".join(dag.predecessors(n))])\n",
    "summary = pd.DataFrame(summary, columns=[\"ID\", \"A\", \"B\", \"C\", \"D\", \"Children\", \"Parents\"])\n",
    "\n",
    "# Save this summary\n",
    "summary.to_csv(\"{0}/MedGen_Term_Enrichments.txt\".format(output_dir), sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a8145a2828964010859e864f85b6ee9a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=9179), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sdw95/.conda/envs/jp/lib/python2.7/site-packages/ipykernel/__main__.py:24: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/home/sdw95/.conda/envs/jp/lib/python2.7/site-packages/ipykernel/__main__.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/home/sdw95/.conda/envs/jp/lib/python2.7/site-packages/ipykernel/__main__.py:26: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/home/sdw95/.conda/envs/jp/lib/python2.7/site-packages/ipykernel/__main__.py:27: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "# Only retain information for the terms that have at least\n",
    "# one gene associated with them\n",
    "summary = summary[(summary[\"A\"] != 0)|(summary[\"C\"] != 0)]\n",
    "\n",
    "# Calculate Odds Ratio for enrichment of SARS2 interactors\n",
    "# within each disease term\n",
    "def do(a, b, c, d):\n",
    "    if([0] in [a, b, c, d]):\n",
    "        return np.nan, np.nan, np.nan, np.nan\n",
    "    exposure_mask = np.zeros(a + b + c + d)\n",
    "    case_mask = np.zeros(a + b + c + d)\n",
    "    exposure_mask[:a+b] = 1\n",
    "    case_mask[:a] = 1\n",
    "    case_mask[a+b:a+b+c] = 1\n",
    "\n",
    "    OR, up, low, p = odds_ratio(exposure_mask, case_mask, log_odds=True, two_sided=False)\n",
    "    pbar.update()\n",
    "    return OR, low, up, p\n",
    "# FUNCTION END\n",
    "\n",
    "# Calculate OR / Format the results into new columns\n",
    "pbar = tqdm_notebook(total=len(summary))\n",
    "tmp = summary[[\"A\", \"B\", \"C\", \"D\"]].apply(lambda x: do(*x), axis=1)\n",
    "summary[\"LogOR\"] = [x[0] for x in tmp]\n",
    "summary[\"Low\"] = [x[1] for x in tmp]\n",
    "summary[\"Up\"] = [x[2] for x in tmp]\n",
    "summary[\"p-value\"] = [x[3] for x in tmp]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n"
     ]
    }
   ],
   "source": [
    "# Number of significnatly enriched terms\n",
    "print len(summary[summary[\"p-value\"] <= 0.05])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CUI</th>\n",
       "      <th>TS</th>\n",
       "      <th>STT</th>\n",
       "      <th>ISPREF</th>\n",
       "      <th>AUI</th>\n",
       "      <th>SAUI</th>\n",
       "      <th>SCUI</th>\n",
       "      <th>SDUI</th>\n",
       "      <th>SAB</th>\n",
       "      <th>TTY</th>\n",
       "      <th>CODE</th>\n",
       "      <th>STR</th>\n",
       "      <th>SUPPRESS</th>\n",
       "      <th>Unnamed: 13</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C0000005</td>\n",
       "      <td>P</td>\n",
       "      <td>PF</td>\n",
       "      <td>Y</td>\n",
       "      <td>A26634265</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M0019694</td>\n",
       "      <td>D012711</td>\n",
       "      <td>MSH</td>\n",
       "      <td>PEP</td>\n",
       "      <td>D012711</td>\n",
       "      <td>(131)I-Macroaggregated Albumin</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C0000005</td>\n",
       "      <td>S</td>\n",
       "      <td>PF</td>\n",
       "      <td>Y</td>\n",
       "      <td>A26634266</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M0019694</td>\n",
       "      <td>D012711</td>\n",
       "      <td>MSH</td>\n",
       "      <td>ET</td>\n",
       "      <td>D012711</td>\n",
       "      <td>(131)I-MAA</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C0000039</td>\n",
       "      <td>P</td>\n",
       "      <td>VC</td>\n",
       "      <td>Y</td>\n",
       "      <td>A0016515</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M0023172</td>\n",
       "      <td>D015060</td>\n",
       "      <td>MSH</td>\n",
       "      <td>MH</td>\n",
       "      <td>D015060</td>\n",
       "      <td>1,2-Dipalmitoylphosphatidylcholine</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C0000039</td>\n",
       "      <td>S</td>\n",
       "      <td>VW</td>\n",
       "      <td>Y</td>\n",
       "      <td>A0100864</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M0023172</td>\n",
       "      <td>D015060</td>\n",
       "      <td>MSH</td>\n",
       "      <td>PM</td>\n",
       "      <td>D015060</td>\n",
       "      <td>Phosphatidylcholine, Dipalmitoyl</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C0000039</td>\n",
       "      <td>S</td>\n",
       "      <td>PF</td>\n",
       "      <td>Y</td>\n",
       "      <td>A0528280</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M0023172</td>\n",
       "      <td>D015060</td>\n",
       "      <td>MSH</td>\n",
       "      <td>N1</td>\n",
       "      <td>D015060</td>\n",
       "      <td>3,5,9-Trioxa-4-phosphapentacosan-1-aminium, 4-...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       #CUI TS STT ISPREF        AUI  SAUI      SCUI     SDUI  SAB  TTY  \\\n",
       "0  C0000005  P  PF      Y  A26634265   NaN  M0019694  D012711  MSH  PEP   \n",
       "1  C0000005  S  PF      Y  A26634266   NaN  M0019694  D012711  MSH   ET   \n",
       "2  C0000039  P  VC      Y   A0016515   NaN  M0023172  D015060  MSH   MH   \n",
       "3  C0000039  S  VW      Y   A0100864   NaN  M0023172  D015060  MSH   PM   \n",
       "4  C0000039  S  PF      Y   A0528280   NaN  M0023172  D015060  MSH   N1   \n",
       "\n",
       "      CODE                                                STR SUPPRESS  \\\n",
       "0  D012711                     (131)I-Macroaggregated Albumin        N   \n",
       "1  D012711                                         (131)I-MAA        N   \n",
       "2  D015060                 1,2-Dipalmitoylphosphatidylcholine        N   \n",
       "3  D015060                   Phosphatidylcholine, Dipalmitoyl        N   \n",
       "4  D015060  3,5,9-Trioxa-4-phosphapentacosan-1-aminium, 4-...        N   \n",
       "\n",
       "   Unnamed: 13  \n",
       "0          NaN  \n",
       "1          NaN  \n",
       "2          NaN  \n",
       "3          NaN  \n",
       "4          NaN  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r",
      "          \r",
      "9179/|/100%|| 9179/9179 [01:50<00:00, 96.10it/s]"
     ]
    }
   ],
   "source": [
    "# Download and Read in Raw MedGen Term Descriptions to get named for each term\n",
    "# NOTE: these files too large to include in repository\n",
    "if(not os.path.exists(\"{0}/MGCONSO.RRF\".format(resource_dir))):\n",
    "    os.system(\"wget https://ftp.ncbi.nlm.nih.gov/pub/medgen/MGCONSO.RRF.gz --directory-prefix {0}\".format(resource_dir))\n",
    "    os.system(\"gunzip {0}/MGREL.RRF.gz\".format(resource_dir))\n",
    "medgen_names = pd.read_csv(\"{0}/MGCONSO.RRF\".format(resource_dir), sep=\"|\")\n",
    "medgen_names.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create term --> name mapping\n",
    "term2name = defaultdict(str)\n",
    "term2name.update(medgen_names[(medgen_names[\"TS\"] == \"P\")&(medgen_names[\"STT\"] == \"PF\")].drop_duplicates(\"#CUI\")[[\"#CUI\", \"STR\"]].set_index(\"#CUI\")[\"STR\"].to_dict())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sdw95/.conda/envs/jp/lib/python2.7/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  from ipykernel import kernelapp as app\n"
     ]
    }
   ],
   "source": [
    "# Update Summary with correct Name\n",
    "summary[\"Name\"] = summary[\"ID\"].map(lambda x: term2name[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add Enrichment information / name back into the network\n",
    "for id, a, b, c, d, OR, low, up, p, name in summary[[\"ID\", \"A\", \"B\", \"C\", \"D\", \"LogOR\", \"Low\", \"Up\", \"p-value\", \"Name\"]].values:\n",
    "    dag.node[id][\"OR\"] = OR\n",
    "    dag.node[id][\"low\"] = low\n",
    "    dag.node[id][\"up\"] = up\n",
    "    dag.node[id][\"p\"] = p\n",
    "    dag.node[id][\"name\"] = name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to add a node and its ancestors from the original dag to a new graph g\n",
    "def add_node_plus_acnestors(g, n, nodes=set(), edges=set()):\n",
    "    # If the node is already in the set of nodes stop\n",
    "    if(n in nodes):\n",
    "        return\n",
    "    \n",
    "    # Add the node to the graph and add it to the list of nodes\n",
    "    g.add_node(n, **dag.node[n])\n",
    "    nodes.add(n)\n",
    "    \n",
    "    # If there are no ancestor nodes for this node,\n",
    "    # add connect it to a universal root node\n",
    "    if(len(list(dag.predecessors(n))) == 0):\n",
    "        g.add_edge(\"root\", n)\n",
    "        edges.add((\"root\", n))\n",
    "    \n",
    "    # Otherwise add all the ancestors\n",
    "    for p in dag.predecessors(n):\n",
    "        if(not (p, n) in edges):\n",
    "            g.add_edge(p, n)\n",
    "            edges.add((p, n))\n",
    "        \n",
    "        add_node_plus_acnestors(g, p, nodes, edges)\n",
    "    \n",
    "    return\n",
    "# FUNCTION END"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c0f039102b9e49b0bcd6c307eba043d2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=2), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Create a sub-set for the term network containing only significant terms and their\n",
    "# ancestors\n",
    "sub_dag = nx.DiGraph()\n",
    "sub_dag.add_node(\"root\", **{\"name\":\"root\"})\n",
    "nodes = set(\"root\")\n",
    "edges = set()\n",
    "for n in tqdm_notebook(summary[summary[\"p-value\"] <= 0.05].sort_values(\"A\", ascending=False)[\"ID\"]):\n",
    "    add_node_plus_acnestors(sub_dag, n, nodes, edges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sub_dag.nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(edges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['#EEEEEE', (0.5, 0.0, 0.0, 1.0), (0.5, 0.0, 0.0, 1.0)]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABIEAAAF1CAYAAAB/DfppAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAIABJREFUeJzs3XmYzXXj//HXmdW+K0S0yDmzYWJkDGOrEAoh4bZkK2afRiJ3hKIZs5VbSbTeEr4iKne2YezMMMs5KZEiJdtYxmzn/P64v1+/u/u+KzLjM3PO83FdrgvD9DxcV8685v0+Y3I4HA4BAAAAAADAqbkZHQAAAAAAAICyxwgEAAAAAADgAhiBAAAAAAAAXAAjEAAAAAAAgAtgBAIAAAAAAHABjEAAAAAAAAAugBEIAAAAAADABTACAQCAW27kyJGaNm2a0Rnlmslk0jfffGN0BgAAcCKMQAAAANepWbNm+vLLL43OAAAA+FMYgQAAQLlRXFxsdILh+DMAAABlhREIAACUuYyMDAUGBqp69eoaPHiwrl69KknasmWLGjdurLlz56pBgwYaNWqUzp07p969e6t+/fqqXbu2evfurR9++EGStHnzZvn7+197v927d1dQUNC1H4eEhGj16tW/29KsWTPFx8crICBANWvW/FWPJH366adq1aqVatWqpeDgYB06dEiSNHz4cB0/flx9+vRRtWrVNG/ePI0YMUIJCQmSpBMnTshkMmnBggWSpG+++UZ16tSRw+GQJC1atEj33nuv6tSpo759++rkyZPX/psmk0mvv/66mjdvrubNm/9H8/bt29WkSRNt3rxZO3bsUL169fT9999Lkg4ePKhatWrJZrNd598GAABwVYxAAACgTBUWFuqxxx7T8OHDdfbsWQ0cOFArV6689vZTp07p7Nmz+u677/Tmm2/Kbrdr1KhR+u6773T8+HFVrlxZkyZNkiS1b99e33zzjX755RcVFxcrOztbP/zwgy5evKj8/Hzt379fHTt2/MOm5cuX6/PPP9fRo0d16NAhLV26VJJ04MABjR49Wm+88YbOnDmj8ePHq2/fviooKNB7772nO++8U2vXrtWlS5cUFxen0NBQbdmyRZK0detW3X333dq6daskKS0tTR07dpTJZNKmTZs0ZcoULV++XD/++KOaNm2qJ5544ldNq1ev1u7du5Wbm/urn//iiy80ZMgQrVy5Ul26dFFwcLDGjx+vESNGKD8/X8OHD9esWbNkNpv/7F8RAABwEYxAAACgTO3atUtFRUWKjIyUp6enHn/8cbVt2/ba293c3DRjxgx5e3urcuXKqlu3rgYMGKAqVaqoevXqmjp16rVhpVKlSmrTpo3S0tK0b98+BQQEKCQkROnp6dq1a5eaN2+uunXr/mFTeHi4GjVqpDp16qhPnz7KzMyU9M/TOuPHj1e7du3k7u6uESNGyNvbW7t27fqv7yc0NFTbtm2T3W5XWlqa4uLilJ6eLumfo1BoaKgk6YMPPtDo0aMVGBgob29vvfzyy9q5c6eOHTt27X1NmTJFderUUeXKla/93Mcff6xx48Zp/fr1vzrx9OKLL+rChQsKCgpSo0aNNHHixOv82wAAAK6MEQgAAJSpkydP6o477pDJZLr2c02bNr32/fr166tSpUrXfnzlyhWNHz9eTZs2VY0aNdSpUyedP39eJSUlknTt9E1aWppCQ0PVuXNnbd269Vejyx9p0KDBte9XqVJFly5dkiR99913SkhIUK1ata59+/777391detf3XPPPapWrZoyMzO1bds29e7dW40aNdJXX331q56TJ0/+6jFXq1ZNdevW1YkTJ679XJMmTf7j/SclJWnQoEG/ugInSZ6enho5cqSys7MVExPzqz9bAACA38IIBAAAylTDhg114sSJa6+NI0nHjx+/9v1/HzASEhL01Vdfaffu3crLy1NaWpokXfv9/z4ChYaG3vAI9FuaNGmiqVOn6vz589e+XblyRUOGDPmvrf/Xs2LFChUWFuqOO+5QaGio3n33XZ07d06tWrWSJDVq1Ejffffdtd9z+fJlnTlzRnfcccdv/jlI/zwJtHr1aiUlJf3q50+cOKEZM2Zo1KhRiomJUUFBwU09bgAA4BoYgQAAQJlq3769PDw8lJKSouLiYq1atUp79uz5zV9/8eJFVa5cWbVq1dLZs2c1Y8aMX709ODhYX331lfbs2aOgoCD5+vrqu+++0+7du9WpU6ebah07dqwWLlyo3bt3y+Fw6PLly1q3bp0uXrwoSbr99tv17bff/ur3hIaG6rXXXrv23+7cubNSU1MVEhIid3d3SdKTTz6pJUuWKDMzUwUFBXr++efVrl07NWvW7Hd7GjVqpI0bNyolJeXaC047HA6NHDlSTz31lBYvXqyGDRvqhRdeuKnHDQAAXAMjEAAAKFNeXl5atWqVli5dqtq1a+ujjz5S//79f/PXR0ZGKj8/X/Xq1dMDDzygHj16/OrtVatWVWBgoHx9feXl5SXpn0NT06ZNddttt91Ua5s2bbRo0SJNmjRJtWvX1r333nvtRaOlf75uz6xZs1SrVi3Fx8dL+ucIdPHixWsjUEhIiK5cufKrQapbt2566aWXNGDAADVs2FBHjhzRsmXLrqvpzjvv1MaNGzV37ly99dZbSklJ0U8//aSXXnpJJpNJS5Ys0ZIlS7Rt27abeuwAAMD5mRz/ejYbAAAAAAAATomTQAAAAAAAAC6AEQgAADiV48ePq1q1av/127++IDUAAICr4ToYAAAAAACAC+AkEAAAAAAAgAtgBAIAAAAAAHABjEAAAAAAAAAugBEIAAAAAADABTACAQAAAAAAuABGIAAAAAAAABfACAQAAAAAAOACGIEAAAAAAABcACMQAAAAAACAC2AEAgAAAAAAcAGMQAAAAAAAAC6AEQgAAAAAAMAFMAIBAAAAAAC4AEYgAAAAAAAAF8AIBAAAAAAA4AIYgQAAAAAAAFwAIxAAAAAAAIALYAQCAAAAAABwAYxAAAAAAAAALsDD6AAAAAAjFRYWKi8vT4WFhfLy8lKNGjXk5eVldBYAAECpYwQCAAAu6eTJk9qxY4e++uorubm5yWQyyeFwyG63y2KxqH379mrYsKHRmQAAAKXG5HA4HEZHAAAA3CqXLl3Shx9+qF9++UXFxcX6b0+FTCaTPDw8VL9+fT355JOqWrWqAaUAAAClixEIAAC4jPPnz+utt95Sfn6+7Hb7H/56Nzc3ValSRWPGjFHNmjVvQSEAAEDZ4YWhAQCAS7h69aqWLl2qK1euXNcAJEl2u12XL1/WO++8o4KCgjIuBAAAKFuMQAAAwCXs3btXly9f/q/Xv36Pw+HQ1gtbFbQwqIzKAAAAbg1GIAAA4PTsdrvSd6aruLj4T/3+EnuJ8vLybnhAAgAAKE94TSAAAOC0miU109Ntntbb+97WkQtHNE7jtF7rdUqnVEM11E3dZJZZknRVV7Ve6/WNvpGnPBWoQHVUR53RGS3UQtllV2WPyvJ099T5584b/MgAAABuHCMQAABwWs2SmqlWpVoKrx+ub7K/0UItVGu1VrCCdVzHtUzLNE7jVE/1tEqrVKAC9Vd/XdEVvaf3FKIQBSpQGcrQAR3QwrYL1atXL6MfFgAAwJ/CdTAAAODUwtuFq3JhZZ3SKRWqUCEKkYc8dLfu1n26T1nKkl125ShH3dRN3vJWbdVWsIJ1UAd/9b4uX75s0KMAAAC4eYxAAADAqTWp0UQmk0kXdVE1VVNu//L0p6Zq6qIu6oquqEQlqqVa//G2f2UymW5ZNwAAQGljBAIAAE7NZDKpevXqqq7quqALsuv/f3n4C7qg6qquKqoiN7npvM7/x9skyaR/jj/Vq1e/tfEAAACliBEIAAA4PV9fX93leZe85KV0patEJTqqozqsw/KTn9zkJl/5apM2qUAFOq/z2qmdClCAJKmqquqiLureFvca/EgAAAD+PA+jAwAAAMpa06ZNVbVSVQ0pGqJ1Wqft2q7qqq5+6qf6qi9J6qVeWq/1SlayPOShQAWqtVpLku7SXWro0VBtP2orN5Obfon7xciHAwAA8Kfw1cEAAIBTczgcWrZsmZYsWaL27dvLze3GD0Lb7XalpaVp8ODBGjt2rNzd3cugFAAAoGxxHQwAADitgwcPKjQ0VPPmzdPUqVPVqFGjGx6B3N3d1bhxYyUkJGjZsmUKDAzUli1byiYYAACgDDECAQAAp3P27FlNnDhRDz74oJ588knt27dPoaGhGj58uGrXrn3dJ3k8PDxUp04dDRs2TIGBgdq8ebOmTZumkSNHauDAgfruu+/K+JEAAACUHkYgAADgNEpKSvTGG2/IYrFIkqxWqyZMmHBt9KlSpYrGjRun5s2by93d/TfHIA8PD7m7u+u+++7TmDFjVLlyZUn//EpjAwcOlNVqlb+/vwIDAzV9+nRdvnz51jxAAACAm8BrAgEAAKeQnp6usLAwVatWTSkpKWrVqtXv/vq8vDzt27dPe/fuVUFBgdzd3VVSUiJvb28FBQWpTZs2f/gl4b///nvFxcUpPT1d8+bN0+DBg2UymUrzYQEAAJQaRiAAAFChnTx5UpMnT9bmzZv16quv6oknnrihIcbhcKi4uFgFBQXy9vaWp6fnDTds27ZNERERqlq1qpKTkxUYGHjD7wMAAKCscR0MAABUSIWFhZo3b54CAgLUuHFj2Ww2DRky5IZP4phMJnl6eqpatWp/agCSpI4dO2rv3r0aMWKEevXqpbFjx+rnn3/+U+8LAACgrDACAQCACufzzz+Xv7+/tm7dqp07d+rll19WtWrVDG1yd3fXmDFjZLPZVKNGDfn6+mr+/PkqLCw0tAsAAOD/cB0MAABUGEeOHFF0dLRyc3OVlJSkRx55xOik32Sz2RQVFaWjR48qMTFRPXv2NDoJAAC4OE4CAQCAcu/y5cuaNm2a2rVrp/bt2ys7O7tcD0CSZDabtX79eiUkJCgiIkK9e/fW4cOHjc4CAAAujBEIAACUWw6HQx999JEsFou+/fZbZWZm6rnnnpO3t7fRadfFZDLpkUceUXZ2tjp37qzg4GA9++yzysvLMzoNAAC4IEYgAABQLmVlZalr166aM2eO3n//fX344Ydq3Lix0Vl/ipeXl2JjY5Wdna0zZ86oRYsWevvtt2W3241OAwAALoQRCAAAlCvnzp1TWFiYunXrpoEDB2r//v3q1KmT0VmlokGDBnr77be1Zs0avfXWWwoKCtKOHTuMzgIAAC6CEQgAAJQLJSUlWrRokSwWi4qKipSbm6tnnnlGHh4eRqeVurZt2yo9PV1RUVEaNGiQhg0bphMnThidBQAAnBwjEAAAMNzOnTvVrl07vfPOO/rss8+0cOFC1atXz+isMmUymTR06FDZbDY1a9ZMLVu21OzZs3X16lWj0wAAgJNiBAIAAIY5deqURowYoccff1yRkZHatm2bWrdubXTWLVWtWjXNmjVLe/bs0YEDB+Tj46NVq1bJ4XAYnQYAAJwMIxAAALjlCgsLFR8fLz8/PzVo0EA2m03Dhg2TyWQyOs0wd999t1auXKlFixbpr3/9q7p3766srCyjswAAgBNhBAIAALfUhg0bFBAQoI0bNyo9PV1z585V9erVjc4qN7p166aMjAz1799f3bp106RJk3TmzBmjswAAgBNgBAIAALfE0aNH1a9fPz399NOaN2+e1q9frxYtWhidVS55eHho4sSJslqtkiSLxaLXX39dxcXFBpcBAICKjBEIAACUqStXrmj69Olq06aN2rRpo5ycHPXt29elr35dr7p16+q1117Txo0btXLlSrVu3VqbNm0yOgsAAFRQjEAAAKBMOBwOrVixQhaLRYcPH1ZmZqamTp2qSpUqGZ1W4fj7+2vjxo2aMWOGnnrqKQ0YMEBHjx41OgsAAFQwjEAAAKDU5eTkqHv37poxY4beeecdLVu2TE2aNDE6q0IzmUzq37+/cnNzFRgYqLZt22ratGm6dOmS0WkAAKCCYAQCAACl5vz584qMjFTnzp312GOPKSMjQ507dzY6y6lUrlxZU6dO1cGDB3Xs2DGZzWZ98MEHfEl5AADwhxiBAADATbPb7Vq8eLHMZrOuXLmi3NxchYWFycPDw+g0p3XHHXfo/fff1/Lly5WYmKgOHTpo3759RmcBAIByzOTg00YAAOAm7N69W2FhYXJ3d1dqaqratGljdJLLsdvtWrp0qaZOnaqePXtqzpw5atCggdFZAACgnOEkEAAA+FN++uknjRo1Sv369dOkSZOUnp7OAGQQNzc3jR49WjabTXXr1pWfn5/i4+NVWFhodBoAAChHGIEAAMANKSoqUmJiovz8/FSvXj3ZbDb95S9/kZsbTyuMVrNmTb366qvasWOHtmzZIj8/P61bt87oLAAAUE5wHQwAAFy3L7/8UuHh4WrSpImSk5NlNpuNTsLv+OyzzxQZGal77rlH8+fP5+8LAAAXx6fsAADAHzp27JgGDBigsWPHas6cOfr8888ZFCqAnj17KisrS927d1fHjh0VExOj8+fPG50FAAAMwggEAAB+U35+vmbMmKH7779frVq1Um5urh577DGZTCaj03CdvLy8FB0drZycHOXl5clsNmvRokUqKSkxOg0AANxiXAcDAAD/weFw6H/+538UExOjtm3b6tVXX1XTpk2NzkIpOHDggMLDw3XlyhWlpKQoJCTE6CQAAHCLMAIBAIBfyc3NVUREhH788UelpKSoa9euRiehlDkcDi1btkxxcXEKCQnRvHnz1KRJE6OzAABAGeM6GAAAkCRduHBB0dHRCg0NVe/evZWRkcEA5KRMJpOGDBkim82m5s2bq1WrVpo5c6by8/ONTgMAAGWIEQgAABdnt9u1dOlSWSwW5eXlKScnRxEREfL09DQ6DWWsatWqmjlzpvbv36+srCxZLBatWLFCHBQHAMA5cR0MAAAXtnfvXoWFhcnhcOi1115T27ZtjU6CgTZv3qyIiAjVrVtXycnJCggIMDoJAACUIk4CAQDggn7++WeNGTNGffv21YQJE7Rz504GIKhLly46cOCABg0apAcffFBPP/20fvnlF6OzAABAKWEEAgDAhRQVFSk5OVm+vr6qUaOGbDabRo4cKTc3nhLgnzw8PPT000/LarXK09NTFotFKSkpKioqMjoNAADcJK6DAQDgIjZt2qTw8HA1aNBAKSkp8vHxMToJFUBOTo4iIyN18uRJJSUl6cEHHzQ6CQAA/EmMQAAAOLnjx48rJiZGe/fu1fz589WvXz+ZTCajs1CBOBwOrVmzRtHR0fL391dCQoLuueceo7MAAMAN4uw3AABO6urVq3rppZfUunVr+fr6Kjc3V/3792cAwg0zmUx69NFHlZOTo3bt2qldu3aaMmWKLl68aHQaAAC4AYxAAAA4GYfDodWrV8vHx0eZmZnav3+/XnzxRVWpUsXoNFRwlSpV0pQpU3To0CGdOHFCZrNZ7777rux2u9FpAADgOnAdDAAAJ2Kz2RQREaHvv/9eKSkp6t69u9FJcGK7du1SeHi43NzclJKSoqCgIKOTAADA7+AkEAAATiAvL0+xsbEKCQlRjx49dPDgQQYglLkHHnhAu3bt0oQJE/TYY49p5MiR+vHHH43OAgAAv4ERCACACsxut+vdd9+V2WzWmTNnlJ2draioKHl6ehqdBhfh5uamkSNHymazqUGDBvL399fcuXNVUFBgdBoAAPg3XAcDAKCC2r9/v8LCwlRcXKzU1FS1a9fO6CRA33zzjWJiYpSTk6P58+erT58+vBg5AADlBCMQAAAVzOnTpzV16lStWbNGs2fP1qhRo+TmxuFelC9ffPGFoqKi1KRJEyUmJsrHx8foJAAAXB7PGAEAqCD+78SPj4+PqlSpIpvNpqeeeooBCOXSww8/rIMHD6pXr14KDQ1VZGSkzp07Z3QWAAAujWeNAABUAFu2bFFgYKD+53/+R5s3b1ZSUpJq1apldBbwuzw9PRUREaHc3FxdvXpVFotFb7zxhkpKSoxOAwDAJXEdDACAcuz7779XbGysdu3apfj4eD3++OO8vgoqrMzMTIWHhysvL08pKSnq1KmT0UkAALgUTgIBAFAOXb16VbNnz1arVq3UokULWa1WDRw4kAEIFVqrVq20detWPf/88xo+fLgGDx6s7777zugsAABcBiMQAADliMPh0Nq1a+Xn56d9+/Zp7969mjlzpqpUqWJ0GlAqTCaTBg0aJKvVKovFosDAQP31r3/VlStXjE4DAMDpcR0MAIBy4vDhw4qIiNDRo0eVkpKihx56yOgkoMwdP35ccXFx2rlzp+bNm6dBgwZx4g0AgDLCSSAAAAx28eJFTZ48WcHBwerevbsOHTrEAASXceedd2rZsmV677339Morryg0NFQZGRlGZwEA4JQYgQAAMIjD4dD7778vs9msU6dOKSsrSzExMfLy8jI6DbjlOnXqpH379mnYsGHq2bOnxo8fr9OnTxudBQCAU2EEAgDAABkZGerYsaMSExP18ccf65133lHDhg2NzgIM5e7urnHjxslqtapKlSry8fFRUlKSioqKjE4DAMApMAIBAHALnTlzRk8//bR69uypESNGaM+ePQoODjY6CyhXateurcTERKWlpemzzz5TQECAvvjiC6OzAACo8BiBAAC4BYqLi7VgwQJZLBZ5enrKarVq7Nixcnd3NzoNKLcsFos+//xzzZs3TxMnTlTfvn319ddfG50FAECFxQgEAEAZS0tLU5s2bbR8+XJt3LhRKSkpql27ttFZQIVgMpnUp08f5eTkKCQkRO3bt1dcXJzy8vKMTgMAoMJhBAIAoIycOHFCTz75pIYOHaopU6Zo8+bN8vf3NzoLqJC8vb0VFxenrKwsnT59WmazWUuWLJHdbjc6DQCACoMRCACAUlZQUKBXXnlFLVu21N133y2bzabBgwfLZDIZnQZUeA0bNtSSJUv0ySef6M0339QDDzygXbt2GZ0FAECFwAgEAEApWrdunfz8/LRjxw7t3r1bs2bNUtWqVY3OApxO27ZtlZ6errCwMA0YMEDDhw/XyZMnjc4CAKBcYwQCAKAUfP311+rdu7eio6OVkpKiNWvW6J577jE6C3Bqbm5uGj58uL766is1adJE/v7+mjNnjq5evWp0GgAA5RIjEAAAN+HSpUuaMmWK2rdvr06dOikrK0s9e/Y0OgtwKdWqVdOcOXO0Z88e7d27V76+vlq9erUcDofRaQAAlCsmB/86AgBwwxwOh5YtW6a4uDh17txZc+fOVaNGjYzOAiDpyy+/VEREhBo2bKikpCT5+fkZnQQAQLnACAQAwA06ePCgwsLCdPHiRb322mvq0KGD0UkA/k1xcbH+9re/aebMmXriiSc0Y8YM1alTx+gsAAAMxXUwAACu09mzZzVx4kQ99NBDGjp0qPbt28cABJRTHh4eCgsLk9VqVUlJiSwWixYsWKDi4mKj0wAAMAwjEAAAf6CkpEQLFy6UxWKRJFmtVo0fP17u7u4GlwH4I/Xq1dOCBQv0j3/8Qx9//LECAwO1efNmo7MAADAE18EAAPgd27dvV1hYmKpXr67U1FS1bNnS6CQAf5LD4dCqVasUGxur+++/X/Hx8WrWrJnRWQAA3DKcBAIA4L84efKkhg0bpieeeEJxcXHaunUrAxBQwZlMJg0YMEC5ublq2bKl7r//fr3wwgu6fPmy0WkAANwSjEAAAPyLwsJCzZs3TwEBAWrSpIlsNpuGDBkik8lkdBqAUlK5cmW98MILyszM1JEjR2Q2m/Xhhx/yJeUBAE6P62AAAPyvzz77TJGRkWrevLkSExPVvHlzo5MA3ALbt29XRESEKleurOTkZN1///1GJwEAUCYYgQAALu/IkSOKioqS1WpVUlKSHnnkEaOTANxiJSUlWrJkiaZNm6bevXtr9uzZuv32243OAgCgVHEdDADgsi5fvqypU6eqXbt2Cg4OVnZ2NgMQ4KLc3d01ZswYffXVV6pZs6Z8fX2VkJCgwsJCo9MAACg1jEAAAJfjcDj00UcfyWKx6NixYzp48KCee+45eXt7G50GwGA1a9ZUQkKC0tPTtXHjRvn7+2v9+vVGZwEAUCq4DgYAcClZWVkKDw/XuXPnlJqaqo4dOxqdBKAcW79+vaKionTvvfdq/vz5atGihdFJAAD8aZwEAgC4hHPnziksLEzdunXTwIEDtW/fPgYgAH+oV69eysrKUteuXdWhQwfFxsbqwoULRmcBAPCnMAIBAJxaSUmJFi1aJIvFouLiYlmtVj3zzDPy8PAwOg1ABeHl5aWYmBjl5OTo3LlzMpvNWrx4sex2u9FpAADcEK6DAQCc1s6dOxUWFqZKlSopNTVVrVu3NjoJgBPYt2+fIiIiVFBQoOTkZHXo0MHoJAAArgsjEADA6Zw6dUqTJ0/Wxo0bNXfuXD355JMymUxGZwFwIg6HQ3//+981efJkderUSXPnzlXjxo2NzgIA4HdxHQwA4DQKCwsVHx8vPz8/NWjQQFarVUOHDmUAAlDqTCaTnnzySVmtVt11111q2bKlZs2apfz8fKPTAAD4TYxAAACnsGHDBgUEBGjjxo1KT0/X3LlzVb16daOzADi5atWqadasWdq3b58yMzPl4+OjlStXisP2AIDyiOtgAIAK7dtvv1V0dLSysrKUlJSk3r17c/IHgGE2bdqkiIgI1a9fX0lJSQoICDA6CQCAazgJBACokK5cuaLp06crKChIQUFBysnJUZ8+fRiAABiqa9euysjI0OOPP67u3btr4sSJOnPmjNFZAABIYgQCAFQwDodDH3/8sSwWiw4fPqyMjAw9//zzqlSpktFpACBJ8vDw0DPPPCObzSY3NzdZLBa99tprKi4uNjoNAODiuA4GAKgwcnJyFB4ertOnTyslJUWdO3c2OgkA/lB2drYiIiL0008/KTk5Wd26dTM6CQDgojgJBAAo986fP6/IyEh16dJF/fr104EDBxiAAFQYfn5++vLLL/XSSy9p7Nix6t+/v7799lujswAALogRCABQbtntdi1evFhms1n5+fnKycnRpEmT5OHhYXQaANwQk8mkfv36KTc3V23atFHbtm31/PMLEKWGAAAgAElEQVTP69KlS0anAQBcCCMQAKBc2r17t9q1a6fFixdr3bp1euONN1S/fn2jswDgplSqVEnPP/+8Dh06pOPHj8tsNuu9996T3W43Og0A4AJ4TSAAQLny008/6bnnntOGDRv0yiuvaOjQoXJz43MWAJzTjh07FBERIQ8PD6WkpKht27ZGJwEAnBjPqgEA5UJRUZESExPl5+enevXqyWq1avjw4QxAAJxacHCwdu/erXHjxunRRx/VqFGjdOrUKaOzAABOimfWAADDffnll2rZsqU+//xzbdu2Ta+++qpq1KhhdBYA3BJubm4aNWqUbDab6tevLz8/P7366qsqKCgwOg0A4GS4DgYAMMyxY8cUExOjjIwMJSYmqm/fvjKZTEZnAYChDh8+rJiYGNlsNiUmJuqRRx7h/40AgFLBSSAAwC2Xn5+vF198Uffff79atWqlnJwcPfroo3yQAwCS7rvvPq1du1apqamKjY1Vz549ZbVajc4CADgBRiAAwC3jcDi0atUqWSwW5ebmKiMjQy+88IIqV65sdBoAlDs9evRQVlaWHn74YXXq1ElRUVE6f/680VkAgAqMEQgAcEvk5ubqoYce0vTp0/X2229r+fLluvPOO43OAoByzdPTU1FRUcrJydHly5dlNpv15ptvqqSkxOg0AEAFxAgEAChTFy5cUHR0tEJDQ9WnTx9lZGSoa9euRmcBQIVy22236c0339T69ev13nvvqU2bNtq2bZvRWQCACoYRCABQJux2u5YsWSKz2ay8vDzl5OQoPDxcnp6eRqcBQIUVGBiotLQ0Pffccxo6dKieeOIJHT9+3OgsAEAFwQgEACh1e/fuVXBwsN544w2tWbNGb731lm677TajswDAKZhMJg0ePFg2m00tWrRQ69atNWPGDF25csXoNABAOccIBAAoNT///LOeeuop9e3bVxMmTNCOHTvUtm1bo7MAwClVqVJFM2bM0IEDB5STkyOLxaLly5fL4XAYnQYAKKcYgQAAN62oqEjJycny9fVVrVq1ZLPZNHLkSLm58c8MAJS1pk2bavny5Xr33Xc1Z84cde7cWZmZmUZnAQDKIZ6dAwBuyqZNm9S6dWt9+umnSktLU0JCgmrWrGl0FgC4nNDQUO3fv19DhgzRww8/rAkTJuj06dNGZwEAyhFGIADAn3L8+HENHDhQo0eP1syZM7VhwwZZLBajswDApbm7u2vChAmy2Wzy9vaWj4+PkpOTVVRUZHQaAKAcYAQCANyQ/Px8zZw5U61bt5afn5+sVqv69+8vk8lkdBoA4H/Vrl1bycnJ2rp1q9atW6eWLVtqw4YNRmcBAAxmcvDKcQCA6+BwOPTJJ58oOjpagYGBio+PV7NmzYzOAgD8AYfDobVr1yo6Olq+vr5KSEjQvffea3QWAMAAnAQCAPwhm82mHj166Pnnn9ebb76pFStWMAABQAVhMpnUt29f5eTkKDg4WA888ICee+45Xbx40eg0AMAtxggEAPhNeXl5io2NVUhIiHr06KGDBw+qe/fuRmcBAP4Eb29vTZ48WYcOHdKPP/4os9msd955R3a73eg0AMAtwggEAPgPdrtd7777rsxms86ePaucnBxFRUXJ09PT6DQAwE1q1KiR3nnnHa1atUoLFixQ+/bttXv3bqOzAAC3AK8JBAD4lf379yssLEzFxcVKTU1Vu3btjE4CAJQRu92u999/X1OmTFG3bt30yiuvqFGjRkZnAQDKCCeBAACSpNOnT2vcuHHq3bu3xowZo127djEAAYCTc3Nz01/+8hfZbDY1atRIAQEBeuWVV3T16lWj0wAAZYARCABc3P+d+PH19VXVqlVltVo1evRoubnxTwQAuIrq1avrlVde0a5du7Rz5075+vrqk08+EZcGAMC5cB0MAFzYli1bFB4ervr16yslJUW+vr5GJwEAyoENGzYoMjJSjRs3VlJSknx8fIxOAgCUAj7NCwAu6Pvvv9fgwYM1YsQITZ8+XV9++SUDEADgmoceekgHDx5U7969FRoaqoiICJ07d87oLADATWIEAgAXcvXqVc2ePVutW7eW2WyW1WrV448/LpPJZHQaAKCc8fT0VHh4uHJzc1VYWCiz2ayFCxeqpKTE6DQAwJ/EdTAAcAEOh0Nr165VVFSUAgICNH/+fN11111GZwEAKpDMzExFRETo/PnzSk5OVufOnY1OAgDcIEYgAHByhw8fVkREhI4dO6bk5GQ99NBDRicBACooh8OhFStWKDY2VkFBQYqPj1fTpk2NzgIAXCeugwGAk7p48aImT56s4OBgde/eXQcPHmQAAgDcFJPJpIEDB8pqtcrf31+BgYGaPn26Ll++bHQaAOA6MAIBgJNxOBx6//33ZTab9dNPPykrK0sxMTHy8vIyOg0A4CSqVKmi6dOnKyMjQ4cPH5bFYtGyZcv4kvIAUM5xHQwAnMiBAwcUFhamgoICpaamqn379kYnAQBcwLZt2xQREaGqVasqOTlZgYGBRicBAP4LTgIBgBM4c+aMJkyYoF69emnUqFHavXs3AxAA4Jbp2LGj9u7dq7/85S/q1auXxo4dq59//tnoLADAv2EEAoAKrLi4WAsWLJDFYpGXl5esVqvGjBkjd3d3o9MAAC7G3d1dY8eOlc1mU/Xq1eXr66vExEQVFhYanQYA+F9cBwOACiotLU3h4eGqXbu2UlJS5O/vb3QSAADXWK1WRUVF6dixY0pKSlKPHj2MTgIAl8cIBAAVzA8//KC4uDht375d8fHxGjhwoEwmk9FZAAD8B4fDoXXr1ikqKkotWrTQ/Pnzdd999xmdBQAui+tgAFBBFBQU6OWXX1arVq109913y2q1atCgQQxAAIByy2QyqXfv3srOzlZoaKiCg4P17LPPKi8vz+g0AHBJjEAAUAF8+umn8vX11a5du7R7927NmjVLVatWNToLAIDr4u3trWeffVbZ2dk6c+aMWrRoobffflt2u93oNABwKVwHA4By7Ouvv1ZkZKS++eYbJScn83oKAACnsHfvXoWHh6uoqEgpKSkKDg42OgkAXAIngQCgHLp06ZKmTJmi9u3bq3PnzsrKymIAAgA4jbZt2yo9PV2RkZEaNGiQhg0bphMnThidBQBOjxEIAMoRh8OhDz/8UGazWT/88IMOHTqkZ599Vl5eXkanAQBQqtzc3DRs2DDZbDY1bdpUAQEBmj17tq5evWp0GgA4La6DAUA5cfDgQYWFhenSpUtKTU1Vhw4djE4CAOCW+fbbbxUbG6vMzEzFx8erX79+fPEDAChljEAAYLAzZ87ohRde0MqVKzVz5kyNGTNG7u7uRmcBAGCIjRs3KiIiQrfffruSkpLk7+9vdBIAOA2ugwGAQUpKSrRw4UL5+PjIzc1NVqtV48ePZwACALi0bt26KTMzU/369VO3bt00adIknTlzxugsAHAKjEAAYIDt27erTZs2+vvf/64NGzbotddeU506dYzOAgCgXPDw8NCkSZNktVrlcDhksVj0+uuvq7i42Og0AKjQuA4GALfQyZMnFRcXp61bt+rVV1/V4MGDeb0DAAD+wKFDhxQREaFffvlFycnJ6tq1q9FJAFAhcRIIAG6BgoICzZ07VwEBAbrzzjtltVr1xBNPMAABAHAdAgICtGnTJs2YMUNPPfWUBgwYoKNHjxqdBQAVDiMQAJSxzz77TP7+/tq+fbt27dqlOXPmqFq1akZnAQBQoZhMJvXv31+5ubkKDAxU27ZtNW3aNF2+fNnoNACoMBiBAKCMHDlyRH369FF4eLiSkpK0du1a3XvvvUZnAQBQoVWuXFlTp05VZmamjh49KrPZrA8++EC8ygUA/DFGIAAoZZcvX9bUqVPVrl07hYSEKDs7W7169TI6CwAAp9K4cWN98MEHWrZsmRITE9WhQwft27fP6CwAKNcYgQCglDgcDn300UeyWCw6duyYDh48qMmTJ8vb29voNAAAnFaHDh20Z88ejRkzRn369NHo0aN16tQpo7MAoFxiBAKAUnDo0CF16dJFL7/8sj744AN98MEHuuOOO4zOAgDAJbi5uWn06NGy2WyqU6eO/Pz8FB8fr8LCQqPTAKBcYQQCgJtw9uxZhYWFqXv37ho8eLD279+vjh07Gp0FAIBLqlmzpuLj45Wenq7NmzfLz89P69atMzoLAMoNRiAA+BNKSkr05ptvymKxqKSkRFarVU8//bTc3d2NTgMAwOW1aNFC69atU1JSkqKjo9WrVy/ZbDajswDAcIxAAHCDdu7cqXbt2um9997T559/rgULFqhu3bpGZwEAgH/Tq1cvZWVlqXv37urYsaNiYmJ04cIFo7MAwDCMQABwnX788UeNGDFCAwcOVHR0tNLS0tS6dWujswAAwO/w8vJSdHS0srOzlZeXJ7PZrLfeekslJSVGpwHALccIBAB/oLCwUPHx8fL391fDhg1ltVr15JNPymQyGZ0GAACu0+23365Fixbp008/1dKlS9W2bVtt377d6CwAuKVMDofDYXQEAJRXX3zxhSIiInT33XcrKSlJ9913n9FJAADgJjkcDi1btkxxcXEKCQnRvHnz1KRJE6OzAKDMcRIIAP6Lb7/9Vo899pgmTpyo+Ph4rVu3jgEIAAAnYTKZNGTIENlsNjVv3lytWrXSzJkzlZ+fb3QaAJQpRiAA+BdXrlzR9OnTFRQUpKCgIGVnZ6t3795c/QIAwAlVrVpVM2fO1P79+3Xo0CFZLBatWLFCXJYA4Ky4DgYA+uex8BUrVig2NlbBwcF69dVX1bhxY6OzAADALbR582ZFRESobt26Sk5OVkBAgNFJAFCqGIEAuLzs7GyFh4frl19+UWpqqkJDQ41OAgAABikuLtaiRYv04osvqn///nrppZdUr149o7MAoFRwHQyAyzp//rwiIiLUtWtX9e/fXwcOHGAAAgDAxXl4eOjpp5+W1WqVp6enfHx8lJKSoqKiIqPTAOCmMQIBcDl2u11vvfWWzGazrl69qtzcXE2aNEkeHh5GpwEAgHKiTp06SklJ0aZNm7RmzRq1atVK//jHP4zOAoCbwnUwAC5l9+7dmjRpkry8vJSamqrAwECjkwAAQDnncDj0ySefKDo6WgEBAUpISNA999xjdBYA3DBOAgFwCT/99JNGjRql/v37Kzw8XNu3b2cAAgAA18VkMumxxx5Tbm6u2rVrp6CgIE2ZMkUXL140Og0AbggjEACnVlRUpPnz58vX11f16tWT1WrV8OHD+ZLvAADghlWqVElTpkzRoUOHdOLECZnNZr377ruy2+1GpwHAdeE6GACn9Y9//EMRERG68847lZSUJLPZbHQSAABwIjt37lRERITc3NyUkpKioKAgo5MA4HcxAgFwOseOHVN0dLQyMzOVmJiovn37cvIHAACUCbvdrnfffVfPP/+8HnroIb388stq2LCh0VkA8F9xHQyA08jPz9eLL76oNm3aKDAwULm5uXr00UcZgAAAQJlxc3PTyJEjZbPZdPvtt8vf319z585VQUGB0WkA8B8YgQBUeA6HQytXrpTFYpHVatWBAwc0bdo0VapUyeg0AADgImrUqKG5c+dq586dSk9Pl6+vr9asWSMuXgAoT7gOBqBCy83NVXh4uH766SelpKSoS5cuRicBAADoiy++UGRkpO68804lJibKx8fH6CQA4CQQgIrpwoULioqKUmhoqPr27auMjAwGIAAAUG48/PDDOnTokHr27KnQ0FBFRkbq3LlzRmcBcHGMQAAqFLvdriVLlshsNuvSpUvKyclReHi4PDw8jE4DAAD4FU9PT0VGRio3N1f5+fkym8164403VFJSYnQaABfFdTAAFcbevXsVFhYmSUpNTVXbtm0NLgIAALh+GRkZioiIUF5enlJSUtSpUyejkwC4GEYgAOXezz//rClTpuizzz7Tyy+/rOHDh8vNjYOMAACg4nE4HFq+fLmeffZZtW/fXvPmzVPTpk2NzgLgIvgoCkC5VVRUpOTkZPn6+qpWrVqyWq0aMWIEAxAAAKiwTCaTBg8eLJvNJovFosDAQL344ou6cuWK0WkAXAAfSQEolzZt2qTWrVvr008/VVpamhISElSzZk2jswAAAEpFlSpV9OKLLyojI0NWq1Vms1kfffQRX1IeQJniOhiAcuX48eOKiYnRvn37NH/+fD322GMymUxGZwEAAJSptLQ0hYeHq0aNGkpOTlbr1q2NTgLghDgJBKBcyM/P18yZMxUYGCh/f3/l5uaqX79+DEAAAMAldOrUSfv379fQoUPVo0cPjR8/XqdPnzY6C4CTYQQCYCiHw6HVq1fL19dXhw4d0v79+zV9+nRVrlzZ6DQAAIBbyt3dXePHj5fNZlPlypXl4+OjpKQkFRUVGZ0GwElwHQyAYWw2myIiIvTDDz8oJSVF3bp1MzoJAACg3MjNzVVUVJSOHz+upKQkPfzww0YnAajgOAkE4JbLy8tTbGysOnbsqJ49eyozM5MBCAAA4N/4+Pjo888/19y5czVx4kT17dtXX3/9tdFZACowRiAAt4zdbtc777wjs9mss2fPKjs7W5GRkfL09DQ6DQAAoFwymUzq27evcnJyFBISovbt22vy5MnKy8szOg1ABcQIBOCW2L9/v0JCQvT6669r9erVevvtt3X77bcbnQUAAFAheHt7Ky4uTllZWfr5559lNpu1dOlS2e12o9MAVCC8JhCAMnX69GlNnTpVa9eu1ezZszVy5Ei5ubE/AwAA3Iw9e/YoPDxcdrtdKSkpeuCBB4xOAlAB8JEYgDJRXFys1NRU+fr6qmrVqrJarRo9ejQDEAAAQCkICgrSjh07FBYWpgEDBmj48OE6efKk0VkAyjk+GgNQ6rZs2aLWrVtr9erV2rx5sxITE1WrVi2jswAAAJyKm5ubhg8fLpvNpiZNmsjf319z5szR1atXjU4DUE5xHQxAqfn+++8VGxur3bt3KyEhQf3795fJZDI6CwAAwCUcOXJEMTExysrKUkJCgh599FGeiwH4FU4CAbhpV69e1ezZs9W6dWuZzWbl5uZqwIABPOkAAAC4he655x6tXr1aCxcu1NSpU/Xggw8qJyfH6CwA5QgjEIA/zeFwaM2aNfL19dX+/fu1d+9ezZgxQ1WqVDE6DQAAwGU9+OCDyszM1KOPPqouXbooLCxMZ8+eNToLQDnACATgT/nqq6/Uq1cvTZ48WQsXLtSqVat01113GZ0FAAAASZ6engoLC1Nubq5KSkpksVj0t7/9TcXFxUanATAQIxCAG3Lx4kXFxcWpQ4cOevDBB3Xo0CE9+OCDRmcBAADgv6hXr54WLFigDRs26KOPPlJgYKA2b95sdBYAgzACAbguDodD77//vsxms37++WdlZ2crOjpanp6eRqcBAADgD7Rs2VKbN2/W9OnTNWrUKD3++OM6duyY0VkAbjFGIAB/6MCBAwoJCVFSUpJWrFihpUuXqkGDBkZnAQAA4AaYTCY9/vjjslqtatmype6//3698MILunz5stFpAG4RRiAAv+mXX37RhAkT1KtXL40aNUp79uxR+/btjc4CAADATahcubJeeOEFZWZm6siRIzKbzfrwww/lcDiMTgNQxhiBAPyH4uJivf766/Lx8ZG3t7esVqvGjBkjNzf+lwEAAOAsmjRpog8//FB///vfFR8fr44dO2r//v1GZwEoQyYHcy+Af5GWlqawsDDVqVNHKSkp8vf3NzoJAAAAZaykpERLlizRtGnT1Lt3b82ZM0e33Xab0VkAShmf1gcgSfrhhx80ZMgQDRs2TFOnTtWmTZsYgAAAAFyEu7u7xowZI5vNppo1a8rHx0cJCQkqLCw0Og1AKWIEAlxcQUGBXn75ZbVq1Ur33nuvrFarBg0aJJPJZHQaAAAAbrFatWopISFB6enp2rhxo/z9/bV+/XqjswCUEq6DAS7s008/VWRkpPz8/DR//nzdfffdRicBAACgHFm3bp2ioqLUvHlzzZ8/Xy1atDA6CcBN4CQQ4IK+/vprPfLII4qJidFrr72m1atXMwABAADgPzzyyCPKzs5Wly5d1KFDB8XGxurChQtGZwH4kxiBABdy6dIlPffcc2rfvr26dOmirKws9ejRw+gsAAAAlGNeXl6KjY1Vdna2zp07J7PZrMWLF8tutxudBuAGMQIBLsDhcOjDDz+U2WzWyZMnlZWVpdjYWHl5eRmdBgAAgAqiQYMGWrx4sdauXavFixcrKChI6enpRmcBuAG8JhDg5A4ePKiwsDBdunRJqamp6tChg9FJAAAAqOD+75OMkydPVmhoqObOnavGjRsbnQXgD3ASCHBSZ86c0TPPPKOHHnpIw4YN0969exmAAAAAUCpMJpOGDh0qm82mu+66Sy1bttSsWbOUn59vdBqA38EIBDiZkpISLVy4UD4+PnJzc5PVatW4cePk7u5udBoAAACcTLVq1TRr1izt27dPGRkZ8vHx0cqVK8WFE6B84joY4ES2b9+usLAw1ahRQ6mpqQoICDA6CQAAAC5k06ZNioiIUP369ZWUlMTzUaCc4SQQ4AROnjypYcOGaciQIZo8ebK2bNnCP7gAAAC45bp27aqMjAwNGDBA3bt318SJE3XmzBmjswD8L0YgoAIrKCjQ3LlzFRAQoKZNm8pqteqJJ56QyWQyOg0AAAAuysPDQxMnTpTVapXJZJLFYvl/7d13dFVV4v7/55Z0EKUKSBGBBJLQAtKbDqAigiiMgwWRUbGEQBCQH6OofFwCUhIQZXB+IuOIIDYcERALohAjPYQURKQJoQmE1Jtbvn8cGYpITXJueb/WYq2Qc8+5T8rBtR/33kevvfaanE6n2dGAgMdyMMBHLVu2TAkJCYqMjNSMGTPUsGFDsyMBAAAAf7B161aNGDFCBw8eVHJysm699VazIwEBixII8DE7duzQyJEjlZ2draSkJN1xxx1mRwIAAAAuyOPx6OOPP9aoUaPUsmVLTZ06VQ0aNDA7FhBwWA4G+Ij8/HyNHz9e7dq1U6dOnbR161YKIAAAAPgEi8Wi/v37KzMzU3FxcWrTpo3Gjx+vvLw8s6MBAYUSCPByHo9HCxcuVFRUlHbt2qUtW7Zo7NixCgkJMTsaAAAAcFlCQ0M1fvx4paWlaffu3YqKitI777wjt9ttdjQgILAcDPBiaWlpGj58uE6cOKFZs2apU6dOZkcCAAAASs3atWuVkJAgu92umTNnqk2bNmZHAvwaM4EAL/Tbb78pPj5ePXr00F//+letX7+eAggAAAB+p0OHDkpNTdVjjz2mu+66S0OGDFFOTo7ZsQC/RQkEeBGXy6W5c+eqSZMmcrlcysjI0BNPPCGbzWZ2NAAAAKBMWK1WDRkyRNnZ2apatapiYmL06quvqri42OxogN9hORjgJdauXav4+HiFh4dr1qxZatGihdmRAAAAgHK3fft2JSYmKjs7WzNmzFDv3r1lsVjMjgX4BUogwGQHDhzQ2LFj9fXXX2vKlCn629/+xn/kAAAAEPCWLVumkSNHqn79+kpKSlJUVJTZkQCfx3IwwCQOh0NTp05VbGysatWqpczMTA0aNIgCCAAAAJB0++23Ky0tTT179lSnTp00cuRIHT9+3OxYgE+jBAJMsGLFCjVr1kzffPONUlJSNGnSJFWsWNHsWAAAAIBXCQ4OVmJiojIyMpSfn6+oqCjNnTtXLpfL7GiAT2I5GFCOdu7cqcTERKWnpyspKUl33nmn2ZEAAAAAn7Fx40YlJCQoLy9PM2fOVOfOnc2OBPgUZgIB5aCgoEDPPfecbr75ZrVt21bbtm2jAAIAAAAuU6tWrbR69WqNHTtW999/v+677z7t2bPH7FiAz6AEAsqQx+PR4sWL1aRJE+3YsUObN2/WuHHjFBISYnY0AAAAwCdZLBbdd999yszMVGRkpFq2bKkXX3xRBQUFZkcDvB7LwYAykp6eruHDh+vIkSOaNWuWunbtanYkAAAAwO/s2rVLY8aMUWpqql599VUNGDCAh60Af4ISCChlx48f14QJE/Tee+9pwoQJevzxx2W3282OBQAAAPi1VatWKSEhQdddd52Sk5PVvHlzsyMBXoflYEApcbvd+te//qWoqCgVFxcrIyNDTz31FAUQAAAAUA66deumDRs26L777lPPnj01bNgwHT582OxYgFehBAJKwQ8//KC2bdtq3rx5+vzzzzVnzhxVrVrV7FgAAABAQLHb7Ro2bJgyMzMVHByspk2bKjk5WSUlJWZHA7wCy8GAq5CTk6Nnn31WK1eu1OTJk3X//fez/hgAAADwEtu2bdOIESP066+/KikpST179jQ7EmAqZgIBV6CkpETTp09XTEyMqlevrqysLD3wwAMUQAAAAIAXiY6O1hdffKFXXnlFTzzxhPr27asdO3aYHQswDSUQcJlWrlyp5s2b64svvtCaNWs0ZcoUVaxY0exYAAAAAM7DYrGob9++ysjIUPv27dWuXTs9++yzOnnypNnRgHJHCQRcol27dql///4aNmyYJk2apGXLlikyMtLsWAAAAAAuQUhIiJ599lmlpaXpwIEDioqK0vz58+V2u82OBpQbSiDgIgoKCjRhwgS1bt1acXFx2rZtm+666y6WfgEAAAA+qFatWpo/f74+/PBDvf7662rfvr1SU1PNjgWUC0og4E94PB59+OGHatq0qbKysrRp0yaNHz9eoaGhZkcDAAAAcJXatWunlJQUPfnkk7r77rs1ePBg7d+/3+xYQJmiBALOIyMjQz169NALL7ygefPmadGiRapTp47ZsQAAAACUIqvVqsGDBys7O1s1a9ZUbGysJk2apKKiIrOjAWWCEgg4w4kTJzRy5Eh17dpVffv21aZNm9S9e3ezYwEAAAAoQxUrVtSkSZOUmpqqlJQURUdHa8mSJfJ4PGZHA0oVJRAgye12a968eYqKilJeXp4yMjIUHx8vu91udjQAAAAA5aRhw4ZasirKMwsAABulSURBVGSJ3njjDY0bN069evVSRkaG2bGAUkMJhID3448/qn379po7d67++9//6s0331S1atXMjgUAAADAJD179tSWLVvUu3dvde3aVQkJCTp27JjZsYCrRgmEgHXw4EENHTpU/fr105NPPqk1a9aodevWZscCAAAA4AWCgoKUkJCgjIwMFRcXKyoqSnPmzJHL5TI7GnDFKIEQcEpKSpSUlKSYmBhdd911ysrK0uDBg2W1cjsAAAAAOFu1atU0Z84crVixQu+9955atWqlb7/91uxYwBWxeNjpCgHk66+/1vDhw1WrVi0lJyerSZMmZkcCAAAA4CM8Ho8WL16s0aNHq23btnr11VdVr149s2MBl4ypDwgIu3fv1oABAzR06FD93//9n1asWEEBBAAAAOCyWCwWDRw4UJmZmYqOjlarVq30/PPPKz8/3+xowCWhBIJfKyws1EsvvaS4uDjFxsYqIyND/fr1k8ViMTsaAAAAAB8VHh6uCRMmaNOmTdq+fbuaNGmihQsX8kh5eD2Wg8EveTweffLJJ0pMTFTr1q01depUpmkCAAAAKBPfffedhg8frgoVKig5OVmtWrUyOxJwXpRA8DtZWVkaPny4fv31V82cOVO33nqr2ZEAAAAA+DmXy6W33npLzz33nPr06aOXX35Z1atXNzsWcBaWg8Fv5Obm6plnnlHnzp11xx13aPPmzRRAAAAAAMqFzWbTo48+qqysLFWoUEHR0dGaMWOGHA6H2dGA/6EEgs9zu92aP3++oqKidOzYMW3btk0jRoxQUFCQ2dEAAAAABJhrr71WM2bM0OrVq7VixQo1a9ZMy5cvNzsWIInlYPBx69evV3x8vNxut2bNmqWbb77Z7EgAAAAAIMnYq3Tp0qUaOXKkIiMjNX36dDVu3NjsWAhgzASCTzp8+LAeffRR9enTR4899phSUlIogAAAAAB4FYvFojvvvFPp6enq0qWLOnTooNGjRys3N9fsaFfF5XDo5P79OrZzpw5t26bDGRk6tnOnTu7fL1dJidnxcAHMBIJPcTqdev311zVx4kQ9+OCDmjBhgipVqmR2LAAAAAC4qJycHI0bN07Lly/Xyy+/rIcfflhWq3fPzXA5HDqUnq79GzZo7/ffa29Kio7v2iWr3S6rzSZZLJIkj9stj9stt9Op6268UXU6dlSdjh1VKy5O1aKjZWO7Dq9ACQSfsWrVKsXHx6tGjRqaOXOmmjZtanYkAAAAALhs69at0/Dhw1VSUqKZM2eqQ4cOZkc6i8fj0S9ffaU1U6Zo16pVCgoLk9vpVElBwWVdJygiQharVc6iIjXo0UMdR49Wva5dZfm9OEL5owTyc26XS8UnTqiksFCu4mJZ7XbZw8IUUrGi7KGhZse7JHv37tUzzzyj1NRUTZs2Tf379+cfDQAAAAA+ze12a8GCBXr22WfVrVs3TZ48WbVr1zY1U9Hx49r01ltKmTZNxbm5cuTlld7FLRYFR0QotHJldRw9Ws0fekgh11xTetfHJaEE8iNup1NHsrK0f8MG7Vu7VnvWrNFvP/0kWSyy2myyWK3yeDzyuFxyO52KuP561W7TRvW6dFHNuDjVbNlSwRUqmP1l/E9RUZGmTp2qpKQkPf300xozZozCw8PNjgUAAAAApSYvL0+vvPKK5syZo8TERI0aNUqh5fw/7Itzc7Vi1Cht/c9/ZLFaL3vGz+UKioiQx+1Wi4cfVo8pU7xqHOrvKIH8wLFfftGPr72mjW++Kf3+47ycxtYWEiJ7SIhKCgt14623quOYMarfrZtps208Ho8+/fRTJSYmqkWLFpo2bZrq169vShYAAAAAKA87d+7UqFGjtGXLFk2bNk39+vUrlzHZz198oY/uv1/FJ0/KVVxc5u93JntoqEIqVdK9Cxeqfrdu5fregYoSyEe5XS79vGKF1kyZol9TU+Vxu+VyOK7+wqem6F17rTqMHq3mgwcrtBw3Xs7OzlZCQoL27Nmj5ORk9ejRo9zeGwAAAADM9uWXX2rEiBGqUaOGkpKSFBsbWybvU5ybq6VPPaWsjz4q85k/F2MPC1PsoEG6LSmJWUFljBLIBx3YtEmLBw5Ufk5O6a7RPEdQeLhksajntGmKe+yxMm2hT548qYkTJ2revHkaN26c4uPjFcTu8QAAAAACkNPp1Jw5c/TSSy9p4MCBevHFF1WlSpVSu/4v33yjDwYOlCMvT86iolK77tU4NSto4AcfqG6nTmbH8Vve/Sw6nMXlcOjLceP0VseOOrZjR5kWQJJUUlCgkvx8fTFqlOZ16qQTe/aU+nt4PB698847ioqK0uHDh7V161YlJiZSAAEAAAAIWHa7XU8//bQyMjLkdrvVpEkTzZ49W06n86qvnb5woRb07q2CI0e8pgCSJGdRkfIPHtQ7PXsq65NPzI7jt5gJ5CMObNyoxQMGKC8nx5Spela7XbaQkFKdFbRx40bFx8fL4XBo1qxZateuXSkkBQAAAAD/kpaWpoSEBB05ckTJycm65ZZbrug66+fM0YrERDkLC0s5Yemyh4Wp9+uvq8XDD5sdxe9QAvmArQsW6NO//90rbtSgiAjd1KuX7l24ULYrnK1z5MgRjR8/XkuWLNHLL7+sIUOGyGplUhoAAAAA/BmPx6OPPvpIzzzzjFq1aqWpU6fqxhtvvOTzN8+bp8+fftr0/X8ulT0sTH3nzVPMX/9qdhS/wsjby62bPdtrCiBJKsnP18/Ll+vd229XyWVmcjqdmj17tpo2barQ0FBlZWVp6NChFEAAAAAAcBEWi0X33HOPMjIy1LJlS7Vu3Vr/+Mc/lJ+ff9FzMz/+WEufespnCiBJchYWasmQIfpp2TKzo/gVZgJ5sfVz5mjFqFFyeuGNag8NVe127fTgihWyBQdf9PWrV69WfHy8KleurFmzZikmJqYcUgIAAACAf9q3b5/Gjh2r1atXa9KkSRo0aNB5t+04un275rRs6ZXjyksRFBGhJ9PTdW39+mZH8QuUQF4qfeFCLXnkEa+ZAXQ+9vBw3fSXv2jgRx/JarOd9zX79u3T6NGjtXbtWk2dOlX33ntvmT5lDAAAAAACyZo1azR8+HCFhoYqOTlZrVu3/t8xt8ulua1b61Bamjxut4kpr5zFZlPtNm30yNq1jCVLAetwvNCR7Gx9OnSoVxdAkuQsKNDOr75SytSpfzhWXFysV155RS1atFCjRo2UkZGhAQMGcNMCAAAAQCnq2LGjfvzxRz3yyCPq06ePHnnkEeXk5EiSfpgxQ7/99JPPFkCS5HG5dHDrVq2fM8fsKH6BmUBexu1yaW6rVjqUnu4zN6o9LEyPb9qkqpGRkqTPPvtMI0aMUExMjKZPn64GDRqYnBAAAAAA/N+JEyc0ceJEvf322xozdKhKZs3y+skFl4plYaWDEsjLrJk8Wd9OnKiSS9jcy1tYrFZVj4lR90WLlDhqlH7++WclJyerV69eZkcDAAAAgICTmZGhee3aKezkSb9Z/sOysNLhL78PfuFIdrZWvfiiTxVAkuRxu5WTman4Vq3UvXt3paWlUQABAAAAgEk8mZm6zuPxqwG/x+XSwfR07Vi+3OwoPs2ffid83scPPCBXcbHZMa6IpaREXTwePTpwoIIv4WlhAAAAAICysWbyZDny8syOUepK8vK0dsoUs2P4NEogL3Fg0yYdzsjwmX2Azsvt1o+vvWZ2CgAAAAAIWIczM3UoPd3sGGVm7w8/6PiuXWbH8FmUQF7ih+nT5fTRWUCnuBwObfjnP33+6wAAAAAAX5WanCx3SYnZMcoOkw+uCiWQFyg6flwZH3wgj8tldpSr5nG7lfnRR2bHAAAAAICA48jPV9o778jtdJodpcy4HA5tmDuXyQdXiBLIC2yeP18Wq3/8KBx5eVozebLZMQAAAAAg4GQsXiz5ydjyYrKXLDE7gk8KjN8OL5c6c6ZKCgrMjlFqjm7frqPbt5sdAwAAAAACys8rV6rEDzeEPpfj5En98tVXZsfwSZRAJnPk5Sl3716zY5Qqq82mfampZscAAAAAgICy74cfzI5QbvasWWN2BJ9ECWSynM2bFRQWZnaMUuXIy9NebkgAAAAAKDfOoiLl7tljdoxy89tPP8ntB/vqljdKIJPt37DBLze0opUFAAAAgPJzcOtWBYWHmx2j3NiCg3UkK8vsGD7HbnaAQLZg6wI9t/8F7R1VrJBi6focqfN3UmiR9EVPaX8tqTBceuGFs887dq20tLe07wbJ5pKaZki3LZdsbuP4Cy9IQY7Tr49Jl/p+anz8TTfpu87Geac88YZU+ZjxcXZj6cu/SMevlWoclO76VKp+2Di2ubmU2lY6WkUKKZZit0q3fnX6fT/sL/1yo+QIlirmb5N9/Vw91vqxUv6uAQAAAADOdWDDBrmdTqXFSintpSNVddY4c2cD6dtu0oD3pegM4xyXVZr4vJSQJF13XFrTQdrcQjpRSQovkNqskzquPf0ebw+WDlWXnHbpumNS92+kqGzjmEfGWHN9a6koVGr0k9Tnv1Lo73MevughpccYx8IKpbgNUpfvTl/7QuNYSdpfU1p+m3SgphTskLr96NDt69erenR0GXw3/RclkEmmp0zXpO8n6c4Uu2r+YJQyOxpK2ZFSq41S9Dbjhlv4tz+eu7S3FJEvjZpm3EDvPCitayO1O2MbnmFzpCq/nf+9o7dJ95znKe5HK0sf3iPd/650wz5pbQfpvb9JT79mFD0lQUbZVPtXqSDcOLa2g9T5e+P8zt9JfZdIdpd0vG6Y/lHr/1NcrTjF1Yq7+m8YAAAAAOBP7V+3Tt82L9D3naQ7P5Ma7jh7nBlUIoUVSN90l5pkSlbPH6/hsUh3f2xMCDh2nTHWvCZXik03jt+2XKp22Bgf7qst/fshKX6WVDFP2tLc+DP0/zcmNnzUX1p2u3T3J8a5rTZK3VZJwSVSbkXj2lWPSE0zT7//n41j88Ol/zxgvH/TDMllk3KvcShn0yZp8OBS/176M5aDmeBE0Qk9/83zmn3HbDXe4lBwiXETRW6Xeq6Uqh6VWm0ybq7zOX6tUeQEOY2breEO6XC1q8+1o6FUb7dUb4+Rp+Ma6WRFaXc943ib9cYxu0u65qQxE2hvndPnVz9sHJMkq9UquT36+djPVx8MAAAAAHBBR08c0jfdpd5LjWLl3HGmdLoYSmt2/mt0WiPVOmCcV/WoFJl99pjv+oOnV4JIRhlzopLx8akJDZVypRCHMZ5Mj5EcQcbxqkeNTKdYPNJvlS/ta0tpb2RvttUYc4Y4pGpHpMJjxy7tAvgfZgKZIGVfioqcRbq7yd2a7Hjoss9vm2rcTPV3GTOBfmok3fL12a+ZN8RocevslXqtMKb2nbK9sTRprFTxpHTzj0a5c4rHcvZ1PBZjul+DX/6YY3e9PxZVn/U2pg86g/LUNKih7mh0x2V/fQAAAACAy5Np3y+nXYq6yDY53b8xllXFbr3w6zyS9tQ1lm2d6d1BxtIyl126aYdUa/8Z55wznnTZjaLn+oPG37/rJK3uIpUES9ce+2OGPxvH7rvBmJ30r6HG9W7YJ93xuVSSn3/hLwJ/QAlkgqMFR1U1vKrsVvsV7WZef5e0sZX0yjjJY5Wabz77Rn94nnFTlARJX98iLRhkTKuzuY0ZRHEbpAp5xo30/kBjql5sutRgp7Ef0C/1jZtuTUej2S0J+mOGTS2MPYvu+vTsz9+51LgZD0ZFKHRkO4XYQi776wMAAAAAXJ4TnkKFF5w9U+d8orKNvXs2tjJm7vyZVd2MQqblprM/f/8CYy+hnQ2MfYdOLStruMMYQ0ZvM/b8+b6T8fkzx5Odv5c6fS/lXC9lRRlj0VMuNI7NvcbYC+ihf0vVD0krexhbmbQ5eMYFcElYDmaCKuFVdKTgiJxup6w222Wd67YYayebZErjX5bGTDZmA63scfo19XcbU+TCiqTblxnLx45UNY5VP2ws5bJ6pLp7jVlFGU2NY9WOSP0+kT6/Q5o2ytj3p9phYw3omTKjjLLogf9IEQV/zGj1SPX323XIdUxvrH/jsr4+AAAAAMDlq2QJU0G4UdBczC1fGzNynH8yLST1ZmN/n0ELTm/5cSabW2q0w9hSJCvS+FzLTcbMnrcflmY/Jd34+2qSc8eTFkk1cyS709if6JQLjWODSowxcO39xrYo3VZJe+tKjvDLG0+DmUCmaH9De4XaQ/VJ1ieyBQfLWXTp7WVhmJRbyVjGZXdJ9kKpxWajKT21zvN8zp2Wd4rFc/ax6IzTO8UXhkqbWp49ve+nhtJ/+xj/GNQ4dOGsbqv082/sCQQAAAAAZa2Jq7bszo3Kijo9pvszN+2UKv9mPGDoXBtbGrN4hswz9ve5ELfV2EBaMiYDdF9l/JGkHTdJFXONbUgudu6fOTVWrXHw/MftEeEXvgD+gJlAJqgUWkkvdX9JT33+lH5qFixHkNHW/tTQeGyeR1KJ3ViKpd8/dv7+cUSBsXZyXRvjnMJQo6G9Psc4fqiadOB6Y8ZQcbC0opfRvJ7auycr0jjHI2M399S2Zy8l21/TODc/XPrsTqnxdmOGkCTtvNHY4X3g+9INv579NeVFSFtjjPd0W6TsOsX676GvdcuNt5TVtxEAAAAA8Lsq11RT92+MlR2ZUfrDOPNct35lLN86U1qs9NWtxrKryufsuXy4qnGtErtx3S3NjH1i6+02jheESb9dZ4w1D1UzxqJdvzXKIbdFWh939lh0XZvTs4UuNo5tsdlYPnbgeuO9V3eR6u6WqlaqUZrfwoDATCCTJLZPVI2IGnr+8NPa20UKdhgzbjp/Z0x7Sx5x+rUv/0OqdFwamWT8/a+LjI281nQ0ZvLc+IuxaZYk5VUwHiGfe40xZa7OXmPWzql1oekx0pK+xrS/a3KNHdtbbDn9Xstulw7WkKy/7x906rqScaMVhUrv3n/6c/V2Sw+8a+RY39oojjwW6doTxZre6w31jepbNt9AAAAAAMD/1GrTRl0WLlSFvAKt7mL8D/wzx5k/33T26+vulWr/Ku1odPpzX99irD6Z+9jpzzVLk/p8Zny8qpvxZGqLR6pyVBqw2HiamGRsJ/Le34ynhUXkG1uPtD5jU+nMJsa2Ii7b6YcUtU01jl1sHNvgF6O0WjDI2DOo7h7pvuXhqjmpVal+DwOBxePxeMwOEchSZ87Ul2PHXtaSMF9QLSZGT269yHbzAAAAAIBS8eu6dXrnL39Rce5F1nD5ieAKFTT0hx9UPTra7Cg+heVgJqsZFydbcLDZMUpd3U6dzI4AAAAAAAGjRmysSgoLzY5RblwOh6pGRZkdw+dQApns+hYt/O5GDa5QQXU6dDA7BgAAAAAEDHtoqCrVrWt2jHJTuVGjy37aNiiBTBccEaFKdeqYHaNUuV0u3dCundkxAAAAACCg3NC2rdkRyg2rT64MJZAXaJuQoKBw/3m0XZXISFVp1OjiLwQAAAAAlJoGPXsquEIFs2OUueCKFXXjLTyJ+kpQAnmB5oMHy+N2mx2jVARXrKhOY8eaHQMAAAAAAk7Te+/1m7HlxUT25UnUV4ISyAuEVqqkpgMGyOIH6xktFoua9O9vdgwAAAAACDjBERFq9tBDstrtZkcpM7bgYLV+/HHZQ0LMjuKTKIG8RPvERJ//JbYFBytu2DC/fNoZAAAAAPiCdiNGyBoUZHaMsmO1qs3TT5udwmdRAnmJ61u0ULXoaFmsvvsjsdhsupmbEQAAAABMUzUyUjViY82OUWbqtG+va+vVMzuGz/LdxsEP9f/Pf2Tz0dlAQRER6vbii373pDMAAAAA8DUdx471yw2igytUUIfRo82O4dMogbxIlcaN1e2FFxQUEWF2lMtisVpVuWFDtU9MNDsKAAAAAAS8yL59VblRI59eaXIui82m6rGxanjbbWZH8Wn+8xvhJ9qPGqXKDRv61M1qCwnRgPffl9UPNrYGAAAAAF9ntdl076JFPrvS5HzsoaG65733ZLFYzI7i03ynaQgQVptNA95/32du1qDwcHV78UVVadzY7CgAAAAAgN9VadRI3SdO9LmVJucTFBGhnlOnshdQKaAE8kJVGjdWv7fflj0szOwoFxQUHq6bevVSh1GjzI4CAAAAADhH+5EjVaVxY59aaXIuq92uGs2aKe7xx82O4hd89zfBz0UPHKjbkpO9tgiyh4Wpdrt2unfRIp/+BwUAAAAA/JXFatW9ixbJHhpqdpQrZgsJYRlYKWL07sXiHn1UvaZP97oiKCg8XHU7ddL9S5fKFhRkdhwAAAAAwJ+o0qiR+r/7rteNKy+FPSxMAz/4gGVgpYgSyMu1HjZMfd96S0Hh4ZIXNJ9BERFqePvtGrR0qU+3yQAAAAAQKKL69dOdb7xhjCt9hD0sTP3efpungZUyi8fj8ZgdAheXs3mzFg8YoJP796ukoKDc399qt8sWEqLbZsxQy7//nal4AAAAAOBjNrz5ppYnJMhZWGh2lAuyh4Wpzz//qWYPPmh2FL9DCeRDXA6Hvn3pJaVMny5nUZFUTj+6oIgI1WzZUv0XLFClOnXK5T0BAAAAAKUvfdEiLRkyxGuLoKDwcN3z3nuKvOsus6P4JUogH5SzebMWDxyovJwcOU6eLLP3CYqIkCwW3TZ9OrN/AAAAAMBP7Fq1SosHDpTj5EljgoEXsIeFKaRSJQ384APV7djR7Dh+ixLIR3ncbv28cqXWTJmifWvXyuN2y+VwXP2FLRYFR0QorEoVdRg9Ws0ffFAh11xz9dcFAAAAAHiN4txcfR4fr8wPPjBly5Ez2cPC1OyBB9RrxgwFR0SYmsXfUQL5geO7d2vd7Nna8M9/yu12yyLJkZd3yefbQ0NlCw5WSWGhburZUx3HjFHdzp2Z+QMAAAAAfm7nl1/qw0GDTJkVdGr2z70LF6p+167l+t6BihLIj7hdLh3dvl0HNmzQ3pQU7fn+ex3NzpY8HllsNlmsVsnjkdvlktvpVMVatVS7bVvV69JFteLiVKN5c1pXAAAAAAgwxbm5Wjl6tLb8+9+yWK1lPjMoKCJCHrdbLYcO1V8mTWIcWo4ogfyc2+WS4+RJlRQWyllUJFtQkOxhYQquUEH2kBCz4wEAAAAAvETRiRPaMn++1r76qoqOH7+sFSYXdcbWIx3HjFGzBx9USMWKpXd9XBJKIAAAAAAAgABgNTsAAAAAAAAAyh4lEAAAAAAAQACgBAIAAAAAAAgAlEAAAAAAAAABgBIIAAAAAAAgAFACAQAAAAAABABKIAAAAAAAgABACQQAAAAAABAAKIEAAAAAAAACACUQAAAAAABAAKAEAgAAAAAACACUQAAAAAAAAAGAEggAAAAAACAAUAIBAAAAAAAEAEogAAAAAACAAEAJBAAAAAAAEAAogQAAAAAAAAIAJRAAAAAAAEAAoAQCAAAAAAAIAJRAAAAAAAAAAYASCAAAAAAAIABQAgEAAAAAAAQASiAAAAAAAIAA8P8AqJiYT5S0QswAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1440x432 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Attempt to plot the sub-network of significant genes\n",
    "#\n",
    "# Can't be meaningfully viewed from here, but the svg can be saved\n",
    "# and inspected\n",
    "\n",
    "import networkx as nx\n",
    "from networkx.drawing.nx_agraph import write_dot, graphviz_layout\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# same layout using matplotlib with no labels\n",
    "#plt.figure(figsize=(200*3, 60*3))\n",
    "plt.figure(figsize=(20, 6))\n",
    "plt.title('draw_networkx')\n",
    "pos = graphviz_layout(sub_dag, prog='dot')\n",
    "\n",
    "\n",
    "# Add color based on level of significance\n",
    "vmax = 10\n",
    "cmap = lambda x: matplotlib.cm.seismic(1 - (float((np.log(x) - np.log(0.05) + vmax)) / (vmax*2.0)))\n",
    "colors = []\n",
    "\n",
    "# Add size based on Odds Ratio\n",
    "sizes = []\n",
    "\n",
    "# Add label from term name\n",
    "labels = dict()\n",
    "for n in sub_dag.nodes:\n",
    "    if(\"p\" in sub_dag.node[n]):\n",
    "        colors.append(cmap(sub_dag.node[n][\"p\"]))\n",
    "        sizes.append(300 + 300*abs(sub_dag.node[n][\"OR\"]))\n",
    "        labels[n] = \"{0} ({1})\\nOR: {2:0.2f} ({3:0.2f}={4:0.2f})\\np: {5}\\n{6}\\t{7}\\n{8}\\t{9}\".format(n, sub_dag.node[n][\"name\"], sub_dag.node[n][\"OR\"], sub_dag.node[n][\"low\"], sub_dag.node[n][\"up\"], sub_dag.node[n][\"p\"], sub_dag.node[n][\"a\"], sub_dag.node[n][\"b\"], sub_dag.node[n][\"c\"], sub_dag.node[n][\"d\"])\n",
    "    else:\n",
    "        colors.append((0.5, 0.5, 0.5, 1))\n",
    "        sizes.append(300)\n",
    "        labels[n] = n\n",
    "\n",
    "nx.draw(sub_dag, pos, labels=None, arrows=False, node_color=colors, node_size=sizes, font_color=\"g\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sdw95/.conda/envs/jp/lib/python2.7/site-packages/ipykernel/__main__.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/home/sdw95/.conda/envs/jp/lib/python2.7/site-packages/ipykernel/__main__.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "# Add list of significant parent / child terms to the summary\n",
    "# This way we can select to view the results based on only the most\n",
    "# specific or most general term (while retaining significance)\n",
    "summary[\"Significant Parents\"] = summary[\"Parents\"].map(lambda x: \",\".join([y for y in x.split(\",\") if y in dag.nodes and \"p\" in dag.node[y] and dag.node[y][\"p\"] <= 0.05]))\n",
    "summary[\"Significant Children\"] = summary[\"Children\"].map(lambda x: \",\".join([y for y in x.split(\",\") if y in dag.nodes and \"p\" in dag.node[y] and dag.node[y][\"p\"] <= 0.05]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cc49c1e058f444aca08811f392082f5e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=9179), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sdw95/.conda/envs/jp/lib/python2.7/site-packages/ipykernel/__main__.py:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6e0d05acba9d43f5a17502909346245c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=9179), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sdw95/.conda/envs/jp/lib/python2.7/site-packages/ipykernel/__main__.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "# Add lists of all Significant Ancestors / Descendents relative to\n",
    "# the given node\n",
    "def do(n, sig=set(), seen=set(), first=True):\n",
    "    if(n in seen):\n",
    "        return sig, seen\n",
    "    seen.add(n)\n",
    "    if(not first and \"p\" in dag.node[n] and dag.node[n][\"p\"] <= 0.05):\n",
    "        sig.add(n)\n",
    "    \n",
    "    for p in dag.predecessors(n):\n",
    "        tmp_sig, tmp_seen = do(p, sig=sig, seen=seen, first=False)\n",
    "        sig.update(tmp_sig)\n",
    "        seen.update(seen)\n",
    "    \n",
    "    if(first):\n",
    "        pbar.update()\n",
    "        return \",\".join(sig)\n",
    "    return sig, seen\n",
    "# FUNCTION END\n",
    "pbar = tqdm_notebook(total=len(summary))\n",
    "tmp = [do(x, sig=set(), seen=set(), first=True) for x in summary[\"ID\"].to_list()]\n",
    "summary[\"Significant Ancestors\"] = tmp\n",
    "\n",
    "def do(n, sig=set(), seen=set(), first=True):\n",
    "    if(n in seen):\n",
    "        return sig, seen\n",
    "    seen.add(n)\n",
    "    if(not first and \"p\" in dag.node[n] and dag.node[n][\"p\"] <= 0.05):\n",
    "        sig.add(n)\n",
    "    \n",
    "    for p in dag.successors(n):\n",
    "        tmp_sig, tmp_seen = do(p, sig=sig, seen=seen, first=False)\n",
    "        sig.update(tmp_sig)\n",
    "        seen.update(seen)\n",
    "    \n",
    "    if(first):\n",
    "        pbar.update()\n",
    "        return \",\".join(sig)\n",
    "    return sig, seen\n",
    "# FUNCTION END\n",
    "pbar = tqdm_notebook(total=len(summary))\n",
    "tmp = [do(x, sig=set(), seen=set(), first=True) for x in summary[\"ID\"].to_list()]\n",
    "summary[\"Significant Descendants\"] = tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the updated Term Enrichments Summary\n",
    "summary.to_csv(\"{0}/MedGen_Term_Enrichments.txt\".format(resource_dir), sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>ID</th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>LogOR</th>\n",
       "      <th>p-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13568</th>\n",
       "      <td></td>\n",
       "      <td>CN239356</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>20339</td>\n",
       "      <td>11.142036</td>\n",
       "      <td>5.323501e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75587</th>\n",
       "      <td>Hyperphosphatasia-intellectual disability synd...</td>\n",
       "      <td>C1855923</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>20335</td>\n",
       "      <td>8.819824</td>\n",
       "      <td>4.673799e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Name        ID  A  B  C  \\\n",
       "13568                                                     CN239356  1  9  1   \n",
       "75587  Hyperphosphatasia-intellectual disability synd...  C1855923  1  9  5   \n",
       "\n",
       "           D      LogOR       p-value  \n",
       "13568  20339  11.142036  5.323501e-08  \n",
       "75587  20335   8.819824  4.673799e-08  "
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# All significant nodes.\n",
    "# Also filtered so that we only so the most general term possible what is still significant\n",
    "# i.e. from any significant node, we continue to go up the network until significance is lost\n",
    "# at the next parent\n",
    "summary[(summary[\"p-value\"] <= 0.05)&(summary[\"Significant Ancestors\"] == \"\")&(summary[\"A\"] >= 1)].sort_values(\"LogOR\", ascending=False)[[\"Name\", \"ID\", \"A\", \"B\", \"C\", \"D\", \"LogOR\", \"p-value\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
